diff -up llvm-project-release-16.x/clang/include/clang/Driver/Distro.h.1~ llvm-project-release-16.x/clang/include/clang/Driver/Distro.h
--- llvm-project-release-16.x/clang/include/clang/Driver/Distro.h.1~	2023-01-28 06:44:11.000000000 +0100
+++ llvm-project-release-16.x/clang/include/clang/Driver/Distro.h	2023-01-30 17:12:18.326183274 +0100
@@ -77,6 +77,8 @@ public:
     UbuntuJammy,
     UbuntuKinetic,
     UbuntuLunar,
+    OpenMandriva,
+    ROSA,
     UnknownDistro
   };
 
@@ -135,6 +137,8 @@ public:
 
   bool IsGentoo() const { return DistroVal == Gentoo; }
 
+  bool IsOpenMandriva() const { return DistroVal == OpenMandriva || DistroVal == ROSA; }
+
   /// @}
 };
 
diff -up llvm-project-release-16.x/clang/lib/Driver/Distro.cpp.1~ llvm-project-release-16.x/clang/lib/Driver/Distro.cpp
--- llvm-project-release-16.x/clang/lib/Driver/Distro.cpp.1~	2023-01-28 06:44:11.000000000 +0100
+++ llvm-project-release-16.x/clang/lib/Driver/Distro.cpp	2023-01-30 17:10:56.420776298 +0100
@@ -33,7 +33,7 @@ static Distro::DistroType DetectOsReleas
   Distro::DistroType Version = Distro::UnknownDistro;
 
   // Obviously this can be improved a lot.
-  for (StringRef Line : Lines)
+  for (StringRef Line : Lines) {
     if (Version == Distro::UnknownDistro && Line.startswith("ID="))
       Version = llvm::StringSwitch<Distro::DistroType>(Line.substr(3))
                     .Case("alpine", Distro::AlpineLinux)
@@ -45,6 +45,15 @@ static Distro::DistroType DetectOsReleas
                     .Case("opensuse", Distro::OpenSUSE)
                     .Case("exherbo", Distro::Exherbo)
                     .Default(Distro::UnknownDistro);
+    else if (Version == Distro::UnknownDistro && Line.startswith("DISTRIB_ID=MandrivaLinux"))
+      // It's one of the Mandriva successors -- will get details
+      // from DISTRIB_DESCRIPTION
+      Version = Distro::OpenMandriva;
+    else if (Version == Distro::OpenMandriva && Line.startswith("DISTRIB_DESCRIPTION=")) {
+      if (Line.substr(20).find("ROSA") != StringRef::npos)
+        Version = Distro::ROSA;
+    }
+  }
   return Version;
 }
 
diff -up llvm-project-release-16.x/clang/lib/Driver/ToolChains/Gnu.cpp.1~ llvm-project-release-16.x/clang/lib/Driver/ToolChains/Gnu.cpp
--- llvm-project-release-16.x/clang/lib/Driver/ToolChains/Gnu.cpp.1~	2023-01-28 06:44:11.000000000 +0100
+++ llvm-project-release-16.x/clang/lib/Driver/ToolChains/Gnu.cpp	2023-01-30 17:14:32.311849376 +0100
@@ -2218,7 +2218,8 @@ void Generic_GCC::GCCInstallationDetecto
   static const char *const AArch64LibDirs[] = {"/lib64", "/lib"};
   static const char *const AArch64Triples[] = {
       "aarch64-none-linux-gnu", "aarch64-linux-gnu", "aarch64-redhat-linux",
-      "aarch64-suse-linux"};
+      "aarch64-suse-linux", "aarch64-openmandriva-linux-gnu",
+      "aarch64-openmandriva-linux-musl", "aarch64-openmandriva-linux-uclibc"};
   static const char *const AArch64beLibDirs[] = {"/lib"};
   static const char *const AArch64beTriples[] = {"aarch64_be-none-linux-gnu",
                                                  "aarch64_be-linux-gnu"};
@@ -2228,7 +2229,28 @@ void Generic_GCC::GCCInstallationDetecto
   static const char *const ARMHFTriples[] = {"arm-linux-gnueabihf",
                                              "armv7hl-redhat-linux-gnueabi",
                                              "armv6hl-suse-linux-gnueabi",
-                                             "armv7hl-suse-linux-gnueabi"};
+                                             "armv7hl-suse-linux-gnueabi",
+                                             "armv7l-openmandriva-linux-gnueabihf",
+                                             "armv7hl-openmandriva-linux-gnueabihf",
+                                             "armv7hnl-openmandriva-linux-gnueabihf",
+                                             "armv8l-openmandriva-linux-gnueabihf",
+                                             "armv8hl-openmandriva-linux-gnueabihf",
+                                             "armv8hnl-openmandriva-linux-gnueabihf",
+                                             "armv8hcnl-openmandriva-linux-gnueabihf",
+                                             "armv7l-openmandriva-linux-musleabihf",
+                                             "armv7hl-openmandriva-linux-musleabihf",
+                                             "armv7hnl-openmandriva-linux-musleabihf",
+                                             "armv8l-openmandriva-linux-musleabihf",
+                                             "armv8hl-openmandriva-linux-musleabihf",
+                                             "armv8hnl-openmandriva-linux-musleabihf",
+                                             "armv8hcnl-openmandriva-linux-musleabihf",
+                                             "armv7l-openmandriva-linux-uclibceabihf",
+                                             "armv7hl-openmandriva-linux-uclibceabihf",
+                                             "armv7hnl-openmandriva-linux-uclibceabihf",
+                                             "armv8l-openmandriva-linux-uclibceabihf",
+                                             "armv8hl-openmandriva-linux-uclibceabihf",
+                                             "armv8hnl-openmandriva-linux-uclibceabihf",
+                                             "armv8hcnl-openmandriva-linux-uclibceabihf"};
   static const char *const ARMebLibDirs[] = {"/lib"};
   static const char *const ARMebTriples[] = {"armeb-linux-gnueabi"};
   static const char *const ARMebHFTriples[] = {
@@ -2248,15 +2270,20 @@ void Generic_GCC::GCCInstallationDetecto
       "x86_64-redhat-linux",    "x86_64-suse-linux",
       "x86_64-manbo-linux-gnu", "x86_64-linux-gnu",
       "x86_64-slackware-linux", "x86_64-unknown-linux",
-      "x86_64-amazon-linux"};
+      "x86_64-amazon-linux",    "x86_64-openmandriva-linux-gnu",
+      "x86_64-openmandriva-linux-musl", "x86_64-openmandriva-linux-uclibc"};
   static const char *const X32Triples[] = {"x86_64-linux-gnux32",
-                                           "x86_64-pc-linux-gnux32"};
+                                           "x86_64-pc-linux-gnux32",
+                                           "x86_64-openmandriva-linux-gnux32",
+                                           "x86_64-openmandriva-linux-muslx32",
+                                           "x86_64-openmandriva-linux-uclibcx32"};
   static const char *const X32LibDirs[] = {"/libx32", "/lib"};
   static const char *const X86LibDirs[] = {"/lib32", "/lib"};
   static const char *const X86Triples[] = {
       "i586-linux-gnu",      "i686-linux-gnu",        "i686-pc-linux-gnu",
       "i386-redhat-linux6E", "i686-redhat-linux",     "i386-redhat-linux",
       "i586-suse-linux",     "i686-montavista-linux", "i686-gnu",
+      "i686-openmandriva-linux-gnu", "i686-openmandriva-linux-musl", "i686-openmandriva-linux-uclibc"
   };
 
   static const char *const LoongArch64LibDirs[] = {"/lib64", "/lib"};
@@ -2310,21 +2337,29 @@ void Generic_GCC::GCCInstallationDetecto
   static const char *const PPC64LibDirs[] = {"/lib64", "/lib"};
   static const char *const PPC64Triples[] = {
       "powerpc64-linux-gnu", "powerpc64-unknown-linux-gnu",
-      "powerpc64-suse-linux", "ppc64-redhat-linux"};
+      "powerpc64-suse-linux", "ppc64-redhat-linux",
+      "ppc64-openmandriva-linux-gnu", "ppc64-openmandriva-linux-musl", "ppc64-openmandriva-linux-uclibc"};
   static const char *const PPC64LELibDirs[] = {"/lib64", "/lib"};
   static const char *const PPC64LETriples[] = {
       "powerpc64le-linux-gnu", "powerpc64le-unknown-linux-gnu",
       "powerpc64le-none-linux-gnu", "powerpc64le-suse-linux",
-      "ppc64le-redhat-linux"};
+      "ppc64le-redhat-linux", "ppc64le-openmandriva-linux-gnu",
+      "ppc64le-openmandriva-linux-musl", "ppc64le-openmandriva-linux-uclibc"};
 
   static const char *const RISCV32LibDirs[] = {"/lib32", "/lib"};
   static const char *const RISCV32Triples[] = {"riscv32-unknown-linux-gnu",
                                                "riscv32-linux-gnu",
-                                               "riscv32-unknown-elf"};
+                                               "riscv32-unknown-elf",
+                                               "riscv32-openmandriva-linux-gnu",
+                                               "riscv32-openmandriva-linux-musl",
+                                               "riscv32-openmandriva-linux-uclibc"};
   static const char *const RISCV64LibDirs[] = {"/lib64", "/lib"};
   static const char *const RISCV64Triples[] = {"riscv64-unknown-linux-gnu",
                                                "riscv64-linux-gnu",
-                                               "riscv64-unknown-elf"};
+                                               "riscv64-unknown-elf",
+                                               "riscv64-openmandriva-linux-gnu",
+                                               "riscv64-openmandriva-linux-musl",
+                                               "riscv64-openmandriva-linux-uclibc"};
 
   static const char *const SPARCv8LibDirs[] = {"/lib32", "/lib"};
   static const char *const SPARCv8Triples[] = {"sparc-linux-gnu",
diff -up llvm-project-release-16.x/clang/lib/Driver/ToolChains/Linux.cpp.1~ llvm-project-release-16.x/clang/lib/Driver/ToolChains/Linux.cpp
--- llvm-project-release-16.x/clang/lib/Driver/ToolChains/Linux.cpp.1~	2023-01-28 06:44:11.000000000 +0100
+++ llvm-project-release-16.x/clang/lib/Driver/ToolChains/Linux.cpp	2023-01-30 17:10:56.421776303 +0100
@@ -137,7 +137,7 @@ std::string Linux::getMultiarchTriple(co
   return TargetTriple.str();
 }
 
-static StringRef getOSLibDir(const llvm::Triple &Triple, const ArgList &Args) {
+static StringRef getOSLibDir(const Distro &distro, const llvm::Triple &Triple, const ArgList &Args) {
   if (Triple.isMIPS()) {
     if (Triple.isAndroid()) {
       StringRef CPUName;
@@ -165,14 +165,15 @@ static StringRef getOSLibDir(const llvm:
   // FIXME: This is a bit of a hack. We should really unify this code for
   // reasoning about oslibdir spellings with the lib dir spellings in the
   // GCCInstallationDetector, but that is a more significant refactoring.
-  if (Triple.getArch() == llvm::Triple::x86 || Triple.isPPC32() ||
-      Triple.getArch() == llvm::Triple::sparc)
+  if (!distro.IsOpenMandriva() &&
+      (Triple.getArch() == llvm::Triple::x86 || Triple.isPPC32() ||
+       Triple.getArch() == llvm::Triple::sparc))
     return "lib32";
 
   if (Triple.getArch() == llvm::Triple::x86_64 && Triple.isX32())
     return "libx32";
 
-  if (Triple.getArch() == llvm::Triple::riscv32)
+  if (!distro.IsOpenMandriva() && Triple.getArch() == llvm::Triple::riscv32)
     return "lib32";
 
   return Triple.isArch32Bit() ? "lib" : "lib64";
@@ -197,7 +198,7 @@ Linux::Linux(const Driver &D, const llvm
   }
 
   if (Distro.IsOpenSUSE() || Distro.IsUbuntu() || Distro.IsAlpineLinux() ||
-      Triple.isAndroid()) {
+      Distro.IsOpenMandriva() || Triple.isAndroid()) {
     ExtraOpts.push_back("-z");
     ExtraOpts.push_back("relro");
   }
@@ -258,7 +259,7 @@ Linux::Linux(const Driver &D, const llvm
   // to the link paths.
   path_list &Paths = getFilePaths();
 
-  const std::string OSLibDir = std::string(getOSLibDir(Triple, Args));
+  const std::string OSLibDir = std::string(getOSLibDir(Distro, Triple, Args));
   const std::string MultiarchTriple = getMultiarchTriple(D, Triple, SysRoot);
 
   // mips32: Debian multilib, we use /libo32, while in other case, /lib is
