--- llvm-3.5/lib/CodeGen/GlobalMerge.cpp.EnableGlobalMerge~	2014-06-19 15:03:05.515698194 +0200
+++ llvm-3.5/lib/CodeGen/GlobalMerge.cpp	2014-06-19 15:03:46.846698327 +0200
@@ -72,7 +72,7 @@ using namespace llvm;
 
 #define DEBUG_TYPE "global-merge"
 
-static cl::opt<bool>
+cl::opt<bool>
 EnableGlobalMerge("enable-global-merge", cl::Hidden,
                   cl::desc("Enable global merge pass"),
                   cl::init(true));
