diff -up llvm-project-16.0.3.src/clang/lib/Driver/ToolChains/Gnu.cpp.121~ llvm-project-16.0.3.src/clang/lib/Driver/ToolChains/Gnu.cpp
--- llvm-project-16.0.3.src/clang/lib/Driver/ToolChains/Gnu.cpp.121~	2023-05-04 18:32:45.779276356 +0200
+++ llvm-project-16.0.3.src/clang/lib/Driver/ToolChains/Gnu.cpp	2023-05-04 18:34:15.382994609 +0200
@@ -2388,7 +2388,9 @@ void Generic_GCC::GCCInstallationDetecto
                                                "riscv32-unknown-elf",
                                                "riscv32-openmandriva-linux-gnu",
                                                "riscv32-openmandriva-linux-musl",
-                                               "riscv32-openmandriva-linux-uclibc"};
+                                               "riscv32-openmandriva-linux-uclibc",
+                                               "riscv32-esp-elf",
+                                               "riscv32-esp-unknown-elf"};
   static const char *const RISCV64LibDirs[] = {"/lib64", "/lib"};
   static const char *const RISCV64Triples[] = {"riscv64-unknown-linux-gnu",
                                                "riscv64-linux-gnu",
