From 359c3d933eeacd15bd1a9651f015ac5aa1a65fcb Mon Sep 17 00:00:00 2001
From: Andrei Safronov <safronov@espressif.com>
Date: Wed, 5 Apr 2023 00:59:41 +0300
Subject: [PATCH 126/140] [Xtensa][LLD] add more tests

---
 lld/test/ELF/xtensa-reloc.s | 44 +++++++++++++++++++++++++++++++++++--
 1 file changed, 42 insertions(+), 2 deletions(-)

diff --git a/lld/test/ELF/xtensa-reloc.s b/lld/test/ELF/xtensa-reloc.s
index 7007756aa2a8..611e671e7646 100644
--- a/lld/test/ELF/xtensa-reloc.s
+++ b/lld/test/ELF/xtensa-reloc.s
@@ -1,17 +1,57 @@
 # REQUIRES: xtensa
 # RUN: llvm-mc -filetype=obj -triple=xtensa -mcpu=esp32 %s -o %t.o
-# RUN: ld.lld %t.o --defsym=a=0x2000 --section-start=.CALL=0x1000 --defsym=b=40 -o %t
+# RUN: ld.lld %t.o --defsym=a=0x2000 --section-start=.CALL=0x1000 --defsym=b=0x40 --defsym=c=0x140 --section-start=.BRANCH=0x5000 --defsym=d=0x5010 --section-start=.BR12=0x100 -o %t
 # RUN: llvm-objdump -d --print-imm-hex %t | FileCheck %s
 
+.section .BR12,"ax",@progbits
+# CHECK-LABEL:section .BR12
+# CHECK:      beqz a2, . +64
+# CHECK-NEXT: bnez a3, . +61
+# CHECK-NEXT: bgez a4, . +58
+# CHECK-NEXT: bltz a5, . +55
+  beqz a2, c
+  bnez a3, c
+  bgez a4, c
+  bltz a5, c
+
 .section .CALL,"ax",@progbits
 # CHECK-LABEL: section .CALL:
 # CHECK:      call0 . +4096
 # CHECK-NEXT: call0 . +4096
 # CHECK-NEXT: call0 . +4092
 # CHECK-NEXT: call0 . +4088
-# CHECK-NEXT: call0 . -4068
+# CHECK-NEXT: j     . +4092
+# CHECK-NEXT: j     . +4089
+# CHECK-NEXT: j     . +4086
+# CHECK-NEXT: j     . -4045
+# CHECK-NEXT: j     . -3792
+# CHECK-NEXT: call0 . -4056
+# CHECK-NEXT: call0 . -3804
+# CHECK-NEXT: l32r a3, . -4065
+# CHECK-NEXT: callx0 a3
+# CHECK-NEXT: l32r a4, . -3815
+# CHECK-NEXT: callx0 a4
   call0 a
   call0 a
   call0 a
   call0 a
+  j a
+  j a
+  j a
+  j b
+  j c
   call0 b
+  call0 c
+  l32r a3, b
+  callx0 a3
+  l32r a4, c
+  callx0 a4
+
+.section .BRANCH,"ax",@progbits
+# CHECK-LABEL: section .BRANCH:
+# CHECK:      beq a3, a4, . +16
+# CHECK-NEXT: ball a3, a4, . +13
+# CHECK-NEXT: blt a3, a4, . +10
+  beq a3, a4, d
+  ball a3, a4, d
+  blt a3, a4, d
-- 
2.40.1

