diff -up llvm-project-main/bolt/runtime/CMakeLists.txt.omv~ llvm-project-main/bolt/runtime/CMakeLists.txt
--- llvm-project-main/bolt/runtime/CMakeLists.txt.omv~	2024-01-23 23:22:51.160781884 +0100
+++ llvm-project-main/bolt/runtime/CMakeLists.txt	2024-01-23 23:23:00.660871669 +0100
@@ -31,7 +31,7 @@ set(BOLT_RT_FLAGS
   -fPIC
   -mgeneral-regs-only)
 if (CMAKE_SYSTEM_PROCESSOR MATCHES "x86_64")
-  set(BOLT_RT_FLAGS ${BOLT_RT_FLAGS} "-mno-sse")
+  set(BOLT_RT_FLAGS ${BOLT_RT_FLAGS} -mno-sse -mfpmath=387)
 endif()
 if (CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64")
   check_cxx_compiler_flag("-mno-outline-atomics" CXX_SUPPORTS_OUTLINE_ATOMICS)
