diff -up llvm-project-release-17.x/bolt/tools/driver/CMakeLists.txt.11~ llvm-project-release-17.x/bolt/tools/driver/CMakeLists.txt
--- llvm-project-release-17.x/bolt/tools/driver/CMakeLists.txt.11~	2023-07-28 09:10:13.000000000 +0200
+++ llvm-project-release-17.x/bolt/tools/driver/CMakeLists.txt	2023-07-28 12:51:25.603391940 +0200
@@ -25,6 +25,8 @@ target_link_libraries(llvm-bolt
   LLVMBOLTProfile
   LLVMBOLTRewrite
   LLVMBOLTUtils
+  LLVMX86CodeGen
+  LLVMAArch64CodeGen
   )
 
 add_bolt_tool_symlink(perf2bolt llvm-bolt)
diff -up llvm-project-release-17.x/bolt/tools/heatmap/CMakeLists.txt.11~ llvm-project-release-17.x/bolt/tools/heatmap/CMakeLists.txt
--- llvm-project-release-17.x/bolt/tools/heatmap/CMakeLists.txt.11~	2023-07-28 12:51:25.604391949 +0200
+++ llvm-project-release-17.x/bolt/tools/heatmap/CMakeLists.txt	2023-07-28 12:52:03.884748033 +0200
@@ -15,6 +15,8 @@ target_link_libraries(llvm-bolt-heatmap
   LLVMBOLTProfile
   LLVMBOLTRewrite
   LLVMBOLTUtils
+  LLVMX86CodeGen
+  LLVMAArch64CodeGen
   )
 
 add_dependencies(bolt llvm-bolt-heatmap)
