diff -up llvm-project-llvmorg-21.1.0-rc1/bolt/runtime/CMakeLists.txt.4~ llvm-project-llvmorg-21.1.0-rc1/bolt/runtime/CMakeLists.txt
--- llvm-project-llvmorg-21.1.0-rc1/bolt/runtime/CMakeLists.txt.4~	2025-07-17 21:02:34.000000000 +0200
+++ llvm-project-llvmorg-21.1.0-rc1/bolt/runtime/CMakeLists.txt	2025-07-18 01:36:26.723123221 +0200
@@ -43,7 +43,7 @@ set(BOLT_RT_FLAGS
 )
 if (CMAKE_SYSTEM_PROCESSOR MATCHES "x86_64")
   set(BOLT_RT_FLAGS ${BOLT_RT_FLAGS} 
-    -mno-sse 
+    -mno-sse -mfpmath=387
     -mgeneral-regs-only)
 endif()
 if (CMAKE_SYSTEM_PROCESSOR MATCHES "riscv64")
