diff -up llvm-project-llvmorg-22.1.0-rc3/runtimes/CMakeLists.txt.49~ llvm-project-llvmorg-22.1.0-rc3/runtimes/CMakeLists.txt
--- llvm-project-llvmorg-22.1.0-rc3/runtimes/CMakeLists.txt.49~	2026-02-10 08:57:16.000000000 +0100
+++ llvm-project-llvmorg-22.1.0-rc3/runtimes/CMakeLists.txt	2026-02-11 23:37:30.455574079 +0100
@@ -178,7 +178,7 @@ endif()
 
 if ("${CMAKE_CXX_COMPILER_ID}" MATCHES "Clang")
   llvm_check_compiler_linker_flag(CXX "-nostdlib++" CXX_SUPPORTS_NOSTDLIBXX_FLAG)
-  if (CXX_SUPPORTS_NOSTDLIBXX_FLAG)
+  if (CXX_SUPPORTS_NOSTDLIBXX_FLAG AND (NOT "${CMAKE_C_COMPILER}" MATCHES ".*gcc"))
     set(CMAKE_REQUIRED_FLAGS "${CMAKE_REQUIRED_FLAGS} -nostdlib++")
   endif()
   check_cxx_compiler_flag(-nostdinc++ CXX_SUPPORTS_NOSTDINCXX_FLAG)
