--- llvm-3.5/tools/clang/lib/Driver/Tools.cpp.armhf	2014-05-05 13:42:31.046661612 +0200
+++ llvm-3.5/tools/clang/lib/Driver/Tools.cpp	2014-05-05 13:44:21.643278964 +0200
@@ -561,6 +561,6 @@ StringRef tools::arm::getARMFloatABI(con
       if (FloatABI != "soft" && FloatABI != "softfp" && FloatABI != "hard") {
         D.Diag(diag::err_drv_invalid_mfloat_abi) << A->getAsString(Args);
-        FloatABI = "soft";
+        FloatABI = "hard";
       }
     }
   }
@@ -692,11 +692,11 @@ StringRef tools::arm::getARMFloatABI(con
         break;
       }
       default:
-        // Assume "soft", but warn the user we are guessing.
-        FloatABI = "soft";
+        // Assume "hard", but warn the user we are guessing.
+        FloatABI = "hard";
         if (Triple.getOS() != llvm::Triple::UnknownOS ||
             !Triple.isOSBinFormatMachO())
-          D.Diag(diag::warn_drv_assuming_mfloat_abi_is) << "soft";
+          D.Diag(diag::warn_drv_assuming_mfloat_abi_is) << "hard";
         break;
       }
     }
