diff -up llvm-9.0.0-20190707/clang/CMakeLists.txt.29~ llvm-9.0.0-20190707/clang/CMakeLists.txt
--- llvm-9.0.0-20190707/clang/CMakeLists.txt.29~	2019-07-07 16:29:47.012528828 +0200
+++ llvm-9.0.0-20190707/clang/CMakeLists.txt	2019-07-07 16:30:39.589532038 +0200
@@ -332,7 +332,7 @@ if (NOT LLVM_LINK_LLVM_DYLIB AND CLANG_L
 endif()
 
 # The libdir suffix must exactly match whatever LLVM's configuration used.
-set(CLANG_LIBDIR_SUFFIX "${LLVM_LIBDIR_SUFFIX}")
+set(CLANG_LIBDIR_SUFFIX "${LLVM_LIBDIR_SUFFIX}" CACHE STRING "")
 
 set(CLANG_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
 set(CLANG_BINARY_DIR ${CMAKE_CURRENT_BINARY_DIR})
