--- tools/libclang/Makefile.orig	2013-11-24 12:16:59.090803758 +0100
+++ tools/libclang/Makefile	2013-11-24 12:18:08.983807239 +0100
@@ -30,7 +30,7 @@ include $(CLANG_LEVEL)/Makefile
 
 # Add soname to the library.
 ifeq ($(HOST_OS), $(filter $(HOST_OS), Linux FreeBSD GNU GNU/kFreeBSD))
-        LLVMLibsOptions += -Wl,-soname,lib$(LIBRARYNAME)$(SHLIBEXT)
+        LLVMLibsOptions += -Wl,-soname,lib$(LIBRARYNAME)-$(LLVMVersion)$(SHLIBEXT)
 endif
 
 ##===----------------------------------------------------------------------===##
