diff -up llvm-project-release-12.x/clang/lib/Driver/ToolChains/Arch/ARM.cpp.2~ llvm-project-release-12.x/clang/lib/Driver/ToolChains/Arch/ARM.cpp
--- llvm-project-release-12.x/clang/lib/Driver/ToolChains/Arch/ARM.cpp.2~	2021-02-08 19:12:17.391076935 +0100
+++ llvm-project-release-12.x/clang/lib/Driver/ToolChains/Arch/ARM.cpp	2021-02-08 19:15:30.716941292 +0100
@@ -250,7 +250,7 @@ arm::FloatABI arm::getARMFloatABI(const
                 .Default(FloatABI::Invalid);
       if (ABI == FloatABI::Invalid && !StringRef(A->getValue()).empty()) {
         D.Diag(diag::err_drv_invalid_mfloat_abi) << A->getAsString(Args);
-        ABI = FloatABI::Soft;
+        ABI = FloatABI::Hard;
       }
     }
   }
@@ -260,16 +260,12 @@ arm::FloatABI arm::getARMFloatABI(const
     ABI = arm::getDefaultFloatABI(Triple);
 
   if (ABI == FloatABI::Invalid) {
-    // Assume "soft", but warn the user we are guessing.
-    if (Triple.isOSBinFormatMachO() &&
-        Triple.getSubArch() == llvm::Triple::ARMSubArch_v7em)
-      ABI = FloatABI::Hard;
-    else
-      ABI = FloatABI::Soft;
+    // Assume "hard", but warn the user we are guessing.
+    ABI = FloatABI::Hard;
 
     if (Triple.getOS() != llvm::Triple::UnknownOS ||
         !Triple.isOSBinFormatMachO())
-      D.Diag(diag::warn_drv_assuming_mfloat_abi_is) << "soft";
+      D.Diag(diag::warn_drv_assuming_mfloat_abi_is) << "hard";
   }
 
   assert(ABI != FloatABI::Invalid && "must select an ABI");
