--- llvm-5.0.0.src/clang/lib/Headers/stdint.h.musl12~	2017-04-27 23:49:45.000000000 +0200
+++ llvm-5.0.0.src/clang/lib/Headers/stdint.h	2017-06-15 23:47:54.427058662 +0200
@@ -22,8 +22,6 @@
  *
 \*===----------------------------------------------------------------------===*/
 
-#ifndef __CLANG_STDINT_H
-#define __CLANG_STDINT_H
 
 /* If we're hosted, fall back to the system's stdint.h, which might have
  * additional definitions.
@@ -72,6 +70,8 @@
 # endif
 
 #else
+#ifndef __CLANG_STDINT_H
+#define __CLANG_STDINT_H
 
 /* C99 7.18.1.1 Exact-width integer types.
  * C99 7.18.1.2 Minimum-width integer types.
@@ -700,5 +700,5 @@ typedef __UINTMAX_TYPE__ uintmax_t;
 #define  INTMAX_C(v) __int_c(v,  __INTMAX_C_SUFFIX__)
 #define UINTMAX_C(v) __int_c(v, __UINTMAX_C_SUFFIX__)
 
-#endif /* __STDC_HOSTED__ */
 #endif /* __CLANG_STDINT_H */
+#endif /* __STDC_HOSTED__ */
