--- llvm-3.8.0.src/tools/clang/lib/Driver/ToolChains.cpp.musl7~	2016-01-26 02:40:51.397870122 +0100
+++ llvm-3.8.0.src/tools/clang/lib/Driver/ToolChains.cpp	2016-01-26 03:02:02.105049789 +0100
@@ -2467,6 +2467,10 @@ void Generic_ELF::addClangTargetOptions(
       getTriple().getArch() == llvm::Triple::aarch64_be ||
       (getTriple().getOS() == llvm::Triple::Linux &&
        (!V.isOlderThan(4, 7, 0) || getTriple().isAndroid())) ||
+      getTriple().getEnvironment() == llvm::Triple::Musl ||
+      getTriple().getEnvironment() == llvm::Triple::MuslEABI ||
+      getTriple().getEnvironment() == llvm::Triple::MuslEABIHF ||
+      getTriple().getEnvironment() == llvm::Triple::MuslX32 ||
       getTriple().getOS() == llvm::Triple::NaCl ||
       (getTriple().getVendor() == llvm::Triple::MipsTechnologies &&
        !getTriple().hasEnvironment());
