--- llvm-3.8.0.src/tools/clang/lib/Driver/ToolChains.cpp.musl11~	2016-01-25 20:58:00.184980632 +0100
+++ llvm-3.8.0.src/tools/clang/lib/Driver/ToolChains.cpp	2016-01-25 20:58:00.190939943 +0100
@@ -4185,6 +4185,10 @@ void Linux::AddCudaIncludeArgs(const Arg
   }
 }
 
+ToolChain::RuntimeLibType Linux::GetDefaultRuntimeLibType() const {
+  return ToolChain::RLT_CompilerRT;
+}
+
 bool Linux::isPIEDefault() const { return getSanitizerArgs().requiresPIE(); }
 
 SanitizerMask Linux::getSupportedSanitizers() const {
--- llvm-3.8.0.src/tools/clang/lib/Driver/ToolChains.h.musl11~	2016-01-25 20:58:00.191933161 +0100
+++ llvm-3.8.0.src/tools/clang/lib/Driver/ToolChains.h	2016-01-25 20:59:07.719885560 +0100
@@ -794,6 +794,7 @@ public:
       llvm::opt::ArgStringList &CC1Args) const override;
   void AddCudaIncludeArgs(const llvm::opt::ArgList &DriverArgs,
                           llvm::opt::ArgStringList &CC1Args) const override;
+  RuntimeLibType GetDefaultRuntimeLibType() const override;
   bool isPIEDefault() const override;
   SanitizerMask getSupportedSanitizers() const override;
   void addProfileRTLibs(const llvm::opt::ArgList &Args,
