diff -up llvm-project-release-19.x/clang/lib/Headers/stdint.h.25~ llvm-project-release-19.x/clang/lib/Headers/stdint.h
--- llvm-project-release-19.x/clang/lib/Headers/stdint.h.25~	2024-07-29 13:45:04.000000000 +0200
+++ llvm-project-release-19.x/clang/lib/Headers/stdint.h	2024-07-29 22:12:38.879860794 +0200
@@ -6,14 +6,6 @@
  *
 \*===----------------------------------------------------------------------===*/
 
-#ifndef __CLANG_STDINT_H
-// AIX system headers need stdint.h to be re-enterable while _STD_TYPES_T
-// is defined until an inclusion of it without _STD_TYPES_T occurs, in which
-// case the header guard macro is defined.
-#if !defined(_AIX) || !defined(_STD_TYPES_T) || !defined(__STDC_HOSTED__)
-#define __CLANG_STDINT_H
-#endif
-
 #if defined(__MVS__) && __has_include_next(<stdint.h>)
 #include_next <stdint.h>
 #else
@@ -65,6 +57,8 @@
 # endif
 
 #else
+#ifndef __CLANG_STDINT_H
+#define __CLANG_STDINT_H
 
 /* C99 7.18.1.1 Exact-width integer types.
  * C99 7.18.1.2 Minimum-width integer types.
@@ -950,6 +944,6 @@ typedef __UINTMAX_TYPE__ uintmax_t;
 #define WINT_WIDTH       __WINT_WIDTH__
 #endif
 
-#endif /* __STDC_HOSTED__ */
 #endif /* __MVS__ */
+#endif /* __STDC_HOSTED__ */
 #endif /* __CLANG_STDINT_H */
