--- llvm-3.8.0.src/tools/clang/lib/Driver/Tools.cpp.armhf~	2016-01-25 20:11:08.436069568 +0100
+++ llvm-3.8.0.src/tools/clang/lib/Driver/Tools.cpp	2016-01-25 20:16:06.255147427 +0100
@@ -676,7 +676,7 @@ arm::FloatABI arm::getARMFloatABI(const
                 .Default(FloatABI::Invalid);
       if (ABI == FloatABI::Invalid && !StringRef(A->getValue()).empty()) {
         D.Diag(diag::err_drv_invalid_mfloat_abi) << A->getAsString(Args);
-        ABI = FloatABI::Soft;
+        ABI = FloatABI::Hard;
       }
     }
 
@@ -736,8 +736,8 @@ arm::FloatABI arm::getARMFloatABI(const
         ABI = (SubArch == 7) ? FloatABI::SoftFP : FloatABI::Soft;
         break;
       default:
-        // Assume "soft", but warn the user we are guessing.
-        ABI = FloatABI::Soft;
+        // Assume "hard", but warn the user we are guessing.
+        ABI = FloatABI::Hard;
         if (Triple.getOS() != llvm::Triple::UnknownOS ||
             !Triple.isOSBinFormatMachO())
           D.Diag(diag::warn_drv_assuming_mfloat_abi_is) << "soft";
