diff -up llvm-project-release-16.x/clang/tools/clang-rename/CMakeLists.txt.14~ llvm-project-release-16.x/clang/tools/clang-rename/CMakeLists.txt
--- llvm-project-release-16.x/clang/tools/clang-rename/CMakeLists.txt.14~	2023-01-30 17:41:32.876919703 +0100
+++ llvm-project-release-16.x/clang/tools/clang-rename/CMakeLists.txt	2023-01-30 17:42:24.566177250 +0100
@@ -16,6 +16,7 @@ clang_target_link_libraries(clang-rename
   clangTooling
   clangToolingCore
   clangToolingRefactoring
+  LLVMSupport
   )
 
 install(FILES clang-rename.py
