diff -up llvm-9.0.0-20190313/clang/lib/Headers/inttypes.h.omv~ llvm-9.0.0-20190313/clang/lib/Headers/inttypes.h
--- llvm-9.0.0-20190313/clang/lib/Headers/inttypes.h.omv~	2019-03-13 15:55:18.603867916 +0100
+++ llvm-9.0.0-20190313/clang/lib/Headers/inttypes.h	2019-03-13 16:32:42.147614938 +0100
@@ -28,6 +28,7 @@
 #endif
 
 #include_next <inttypes.h>
+#include "/usr/include/inttypes.h"
 
 #if defined(_MSC_VER) && _MSC_VER < 1900
 /* MSVC headers define int32_t as int, but PRIx32 as "lx" instead of "x".
diff -up llvm-9.0.0-20190313/clang/lib/Headers/limits.h.omv~ llvm-9.0.0-20190313/clang/lib/Headers/limits.h
--- llvm-9.0.0-20190313/clang/lib/Headers/limits.h.omv~	2019-03-13 18:09:07.850876194 +0100
+++ llvm-9.0.0-20190313/clang/lib/Headers/limits.h	2019-03-13 18:09:18.565064422 +0100
@@ -35,6 +35,7 @@
    Include it if we're hosted. */
 #if __STDC_HOSTED__ && __has_include_next(<limits.h>)
 #include_next <limits.h>
+#include "/usr/include/limits.h"
 #endif
 
 /* Many system headers try to "help us out" by defining these.  No really, we
