diff -up llvm-9.0.0-20190313/clang/CMakeLists.txt.24~ llvm-9.0.0-20190313/clang/CMakeLists.txt
--- llvm-9.0.0-20190313/clang/CMakeLists.txt.24~	2019-03-13 02:29:11.218969778 +0100
+++ llvm-9.0.0-20190313/clang/CMakeLists.txt	2019-03-13 02:36:22.398164361 +0100
@@ -305,7 +305,7 @@ set(CLANG_PYTHON_BINDINGS_VERSIONS "" CA
     "Python versions to install libclang python bindings for")
 
 # The libdir suffix must exactly match whatever LLVM's configuration used.
-set(CLANG_LIBDIR_SUFFIX "${LLVM_LIBDIR_SUFFIX}")
+set(CLANG_LIBDIR_SUFFIX "${LLVM_LIBDIR_SUFFIX}" CACHE STRING "")
 
 set(CLANG_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR})
 set(CLANG_BINARY_DIR ${CMAKE_CURRENT_BINARY_DIR})
