diff -up llvm-project-main/clang/lib/Headers/stdint.h.20~ llvm-project-main/clang/lib/Headers/stdint.h
--- llvm-project-main/clang/lib/Headers/stdint.h.20~	2022-01-18 18:01:52.000000000 +0100
+++ llvm-project-main/clang/lib/Headers/stdint.h	2022-01-18 18:40:57.214876195 +0100
@@ -6,14 +6,6 @@
  *
 \*===----------------------------------------------------------------------===*/
 
-#ifndef __CLANG_STDINT_H
-// AIX system headers need stdint.h to be re-enterable while _STD_TYPES_T
-// is defined until an inclusion of it without _STD_TYPES_T occurs, in which
-// case the header guard macro is defined.
-#if !defined(_AIX) || !defined(_STD_TYPES_T) || !defined(__STDC_HOSTED__)
-#define __CLANG_STDINT_H
-#endif
-
 /* If we're hosted, fall back to the system's stdint.h, which might have
  * additional definitions.
  */
@@ -61,6 +53,8 @@
 # endif
 
 #else
+#ifndef __CLANG_STDINT_H
+#define __CLANG_STDINT_H
 
 /* C99 7.18.1.1 Exact-width integer types.
  * C99 7.18.1.2 Minimum-width integer types.
@@ -857,5 +851,5 @@ typedef __UINTMAX_TYPE__ uintmax_t;
 #define WINT_WIDTH       __WINT_WIDTH__
 #endif
 
-#endif /* __STDC_HOSTED__ */
 #endif /* __CLANG_STDINT_H */
+#endif /* __STDC_HOSTED__ */
