diff -up llvm-project-release-15.x/clang/CMakeLists.txt.20~ llvm-project-release-15.x/clang/CMakeLists.txt
--- llvm-project-release-15.x/clang/CMakeLists.txt.20~	2022-07-31 13:12:17.249816158 +0200
+++ llvm-project-release-15.x/clang/CMakeLists.txt	2022-07-31 13:21:00.054974790 +0200
@@ -364,7 +364,7 @@ if (NOT LLVM_LINK_LLVM_DYLIB AND CLANG_L
 endif()
 
 # The libdir suffix must exactly match whatever LLVM's configuration used.
-set(CLANG_LIBDIR_SUFFIX "${LLVM_LIBDIR_SUFFIX}")
+set(CLANG_LIBDIR_SUFFIX "${LLVM_LIBDIR_SUFFIX}" CACHE STRING "")
 
 set(CLANG_TOOLS_INSTALL_DIR "${CMAKE_INSTALL_BINDIR}" CACHE PATH
     "Path for binary subdirectory (defaults to '${CMAKE_INSTALL_BINDIR}')")
