--- llvm-3.7.0.src/tools/clang/lib/Driver/ToolChains.cpp.musl11~	2015-06-27 14:19:27.167903119 +0200
+++ llvm-3.7.0.src/tools/clang/lib/Driver/ToolChains.cpp	2015-06-27 14:20:50.057494250 +0200
@@ -3646,6 +3646,10 @@ void Linux::AddClangCXXStdlibIncludeArgs
   }
 }
 
+ToolChain::RuntimeLibType Linux::GetDefaultRuntimeLibType() const {
+  return ToolChain::RLT_CompilerRT;
+}
+
 bool Linux::isPIEDefault() const { return getSanitizerArgs().requiresPIE(); }
 
 SanitizerMask Linux::getSupportedSanitizers() const {
--- llvm-3.7.0.src/tools/clang/lib/Driver/ToolChains.h.musl11~	2015-06-27 14:19:27.168903114 +0200
+++ llvm-3.7.0.src/tools/clang/lib/Driver/ToolChains.h	2015-06-27 14:21:22.015345035 +0200
@@ -644,6 +644,7 @@ public:
   void AddClangCXXStdlibIncludeArgs(
       const llvm::opt::ArgList &DriverArgs,
       llvm::opt::ArgStringList &CC1Args) const override;
+  RuntimeLibType GetDefaultRuntimeLibType() const override;
   bool isPIEDefault() const override;
   SanitizerMask getSupportedSanitizers() const override;
 
