--- tools/libclang/Makefile.soname~	2014-08-25 22:21:51.899931322 +0200
+++ tools/libclang/Makefile	2014-08-25 22:22:40.725320921 +0200
@@ -34,7 +34,7 @@ include $(CLANG_LEVEL)/Makefile
 
 # Add soname to the library.
 ifeq ($(HOST_OS), $(filter $(HOST_OS), Linux FreeBSD GNU GNU/kFreeBSD))
-        LLVMLibsOptions += -Wl,-soname,lib$(LIBRARYNAME)$(SHLIBEXT)
+        LLVMLibsOptions += -Wl,-soname,lib$(LIBRARYNAME)-$(LLVMVersion)$(SHLIBEXT)
 endif
 
 ifeq ($(ENABLE_CLANG_ARCMT),1)
