diff -up llvm-project-main/compiler-rt/lib/builtins/int_math.h.22~ llvm-project-main/compiler-rt/lib/builtins/int_math.h
--- llvm-project-main/compiler-rt/lib/builtins/int_math.h.22~	2024-01-21 20:44:02.916149920 +0100
+++ llvm-project-main/compiler-rt/lib/builtins/int_math.h	2024-01-21 20:45:48.177116732 +0100
@@ -87,22 +87,8 @@
 #endif
 #endif
 
-#if defined(_MSC_VER) && !defined(__clang__)
-#define crt_fmaxl(x, y) __max((x), (y))
-#else
 #define crt_fmaxl(x, y) __builtin_fmaxl((x), (y))
-#endif
-
-#if defined(_MSC_VER) && !defined(__clang__)
-#define crt_logbl(x) logbl((x))
-#else
 #define crt_logbl(x) __builtin_logbl((x))
-#endif
-
-#if defined(_MSC_VER) && !defined(__clang__)
-#define crt_scalbnl(x, y) scalbnl((x), (y))
-#else
 #define crt_scalbnl(x, y) __builtin_scalbnl((x), (y))
-#endif
 
 #endif // INT_MATH_H
