diff -up llvm-project-release-13.x/compiler-rt/lib/builtins/int_math.h.18~ llvm-project-release-13.x/compiler-rt/lib/builtins/int_math.h
--- llvm-project-release-13.x/compiler-rt/lib/builtins/int_math.h.18~	2021-08-02 22:25:24.032831260 +0200
+++ llvm-project-release-13.x/compiler-rt/lib/builtins/int_math.h	2021-08-02 22:28:26.322670596 +0200
@@ -77,22 +77,10 @@
 #define crt_fabsl(x) __builtin_fabsl((x))
 #endif
 
-#if defined(_MSC_VER) && !defined(__clang__)
-#define crt_fmaxl(x, y) __max((x), (y))
-#else
 #define crt_fmaxl(x, y) __builtin_fmaxl((x), (y))
-#endif
 
-#if defined(_MSC_VER) && !defined(__clang__)
-#define crt_logbl(x) logbl((x))
-#else
 #define crt_logbl(x) __builtin_logbl((x))
-#endif
 
-#if defined(_MSC_VER) && !defined(__clang__)
-#define crt_scalbnl(x, y) scalbnl((x), (y))
-#else
 #define crt_scalbnl(x, y) __builtin_scalbnl((x), (y))
-#endif
 
 #endif // INT_MATH_H
