--- llvm-3.5/projects/compiler-rt/lib/sanitizer_common/sanitizer_platform_limits_posix.cc.omv~	2014-05-05 18:19:40.475765140 +0200
+++ llvm-3.5/projects/compiler-rt/lib/sanitizer_common/sanitizer_platform_limits_posix.cc	2014-05-05 20:22:35.137569180 +0200
@@ -12,7 +12,6 @@
 // Sizes and layouts of platform-specific POSIX data structures.
 //===----------------------------------------------------------------------===//
 
-
 #include "sanitizer_platform.h"
 #if SANITIZER_LINUX || SANITIZER_FREEBSD || SANITIZER_MAC
 
--- llvm-3.5/projects/compiler-rt/lib/sanitizer_common/sanitizer_platform_limits_posix.h.omv~	2014-05-05 18:15:58.681537949 +0200
+++ llvm-3.5/projects/compiler-rt/lib/sanitizer_common/sanitizer_platform_limits_posix.h	2014-05-05 20:22:26.352679006 +0200
@@ -278,7 +278,7 @@ namespace __sanitizer {
   struct __sanitizer_XDR {
     int x_op;
     struct xdr_ops {
-      uptr fns[10];
+      uptr fns[9];
     } *x_ops;
     uptr x_public;
     uptr x_private;
