diff -up llvm-9.0.0-20190618/llvm/include/llvm/ADT/Triple.h.39~ llvm-9.0.0-20190618/llvm/include/llvm/ADT/Triple.h
--- llvm-9.0.0-20190618/llvm/include/llvm/ADT/Triple.h.39~	2019-06-18 22:41:58.457967351 +0200
+++ llvm-9.0.0-20190618/llvm/include/llvm/ADT/Triple.h	2019-06-18 22:42:56.984534475 +0200
@@ -722,6 +722,11 @@ public:
     return getArch() == Triple::ppc64 || getArch() == Triple::ppc64le;
   }
 
+  /// Tests whether the target is RISC-V (32- and 64-bit).
+  bool isRISCV() const {
+    return getArch() == Triple::riscv32 || getArch() == Triple::riscv64;
+  }
+
   /// Tests whether the target supports comdat
   bool supportsCOMDAT() const {
     return !isOSBinFormatMachO();
