--- llvm-3.7.0.src/compiler-rt/lib/tsan/dd/dd_interceptors.cc.omv~	2015-06-02 13:45:40.856457552 +0200
+++ llvm-3.7.0.src/compiler-rt/lib/tsan/dd/dd_interceptors.cc	2015-06-02 13:46:05.401452554 +0200
@@ -7,11 +7,14 @@
 //
 //===----------------------------------------------------------------------===//
 
+// Make sure we can redefine realpath later...
+#define realpath __libc_realpath
+#include <stdlib.h>
+#undef realpath
 #include "dd_rtl.h"
 #include "interception/interception.h"
 #include "sanitizer_common/sanitizer_procmaps.h"
 #include <pthread.h>
-#include <stdlib.h>
 
 using namespace __dsan;
 
