diff -up llvm-project-main/flang/CMakeLists.txt.39~ llvm-project-main/flang/CMakeLists.txt
--- llvm-project-main/flang/CMakeLists.txt.39~	2024-01-21 17:13:12.000000000 +0100
+++ llvm-project-main/flang/CMakeLists.txt	2024-01-21 20:49:43.075284165 +0100
@@ -6,6 +6,12 @@ endif()
 include(${LLVM_COMMON_CMAKE_UTILS}/Modules/CMakePolicy.cmake
   NO_POLICY_SCOPE)
 
+if("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "riscv64")
+  set(atomic_libs "atomic")
+else()
+  set(atomic_libs)
+endif()
+
 set(CMAKE_BUILD_WITH_INSTALL_NAME_DIR ON)
 
 # Flang requires C++17.
diff -up llvm-project-main/flang/tools/bbc/CMakeLists.txt.39~ llvm-project-main/flang/tools/bbc/CMakeLists.txt
--- llvm-project-main/flang/tools/bbc/CMakeLists.txt.39~	2024-01-21 17:13:12.000000000 +0100
+++ llvm-project-main/flang/tools/bbc/CMakeLists.txt	2024-01-21 20:49:43.075284165 +0100
@@ -32,4 +32,6 @@ FortranParser
 FortranEvaluate
 FortranSemantics
 FortranLower
+LLVMSupport
+${atomic_libs}
 )
diff -up llvm-project-main/flang/tools/fir-opt/CMakeLists.txt.39~ llvm-project-main/flang/tools/fir-opt/CMakeLists.txt
--- llvm-project-main/flang/tools/fir-opt/CMakeLists.txt.39~	2024-01-21 20:49:43.075284165 +0100
+++ llvm-project-main/flang/tools/fir-opt/CMakeLists.txt	2024-01-21 20:50:37.347887784 +0100
@@ -36,4 +36,5 @@ target_link_libraries(fir-opt PRIVATE
   MLIRSupport
   MLIRVectorToLLVM
   MLIROptLib
+  ${atomic_libs}
 )
diff -up llvm-project-main/flang/tools/flang-driver/CMakeLists.txt.39~ llvm-project-main/flang/tools/flang-driver/CMakeLists.txt
--- llvm-project-main/flang/tools/flang-driver/CMakeLists.txt.39~	2024-01-21 17:13:12.000000000 +0100
+++ llvm-project-main/flang/tools/flang-driver/CMakeLists.txt	2024-01-21 20:49:43.075284165 +0100
@@ -28,12 +28,14 @@ target_link_libraries(flang-new
   PRIVATE
   flangFrontend
   flangFrontendTool
+  ${atomic_libs}
 )
 
 clang_target_link_libraries(flang-new
   PRIVATE
   clangDriver
   clangBasic
+  ${atomic_libs}
 )
 
 option(FLANG_PLUGIN_SUPPORT "Build Flang with plugin support." ON)
diff -up llvm-project-main/flang/tools/tco/CMakeLists.txt.39~ llvm-project-main/flang/tools/tco/CMakeLists.txt
--- llvm-project-main/flang/tools/tco/CMakeLists.txt.39~	2024-01-21 17:13:12.000000000 +0100
+++ llvm-project-main/flang/tools/tco/CMakeLists.txt	2024-01-21 20:49:43.075284165 +0100
@@ -30,4 +30,5 @@ target_link_libraries(tco PRIVATE
   MLIRParser
   MLIRSupport
   MLIRVectorToLLVM
+  ${atomic_libs}
 )
