diff -up llvm-project-release-19.x/libcxx/include/__config.26~ llvm-project-release-19.x/libcxx/include/__config
--- llvm-project-release-19.x/libcxx/include/__config.26~	2024-07-29 22:13:08.148127223 +0200
+++ llvm-project-release-19.x/libcxx/include/__config	2024-07-29 22:14:25.488853338 +0200
@@ -10,6 +10,13 @@
 #ifndef _LIBCPP___CONFIG
 #define _LIBCPP___CONFIG
 
+#include <features.h>
+#if defined(__linux__) && !defined(__ANDROID__) && !defined(__GLIBC__)
+// If it's Linux, but neither GLIBC nor Android, it's likely musl...
+// (uClibc defined __GLIBC__ for compatibility)
+#define _LIBCPP_HAS_MUSL_LIBC
+#endif
+
 #include <__config_site>
 #include <__configuration/abi.h>
 #include <__configuration/availability.h>
