diff -up llvm-9.0.0-20190618/libunwind/src/UnwindRegistersRestore.S.13~ llvm-9.0.0-20190618/libunwind/src/UnwindRegistersRestore.S
--- llvm-9.0.0-20190618/libunwind/src/UnwindRegistersRestore.S.13~	2019-06-18 20:22:30.000000000 +0200
+++ llvm-9.0.0-20190618/libunwind/src/UnwindRegistersRestore.S	2019-06-18 20:58:37.918780439 +0200
@@ -559,6 +559,10 @@ Lnovec:
 
 #elif defined(__arm64__) || defined(__aarch64__)
 
+// Let's make GNU as happy
+#define fp x29
+#define lr x30
+
 //
 // void libunwind::Registers_arm64::jumpto()
 //
diff -up llvm-9.0.0-20190618/libunwind/src/UnwindRegistersSave.S.13~ llvm-9.0.0-20190618/libunwind/src/UnwindRegistersSave.S
--- llvm-9.0.0-20190618/libunwind/src/UnwindRegistersSave.S.13~	2019-06-18 20:58:37.918780439 +0200
+++ llvm-9.0.0-20190618/libunwind/src/UnwindRegistersSave.S	2019-06-18 20:59:45.981405783 +0200
@@ -699,6 +699,10 @@ DEFINE_LIBUNWIND_FUNCTION(__unw_getconte
 
 #elif defined(__arm64__) || defined(__aarch64__)
 
+// Make GNU as happy
+#define fp x29
+#define lr x30
+
 //
 // extern int __unw_getcontext(unw_context_t* thread_state)
 //
