diff -up llvm-9.0.0-20190729/lld/ELF/Config.h.19~ llvm-9.0.0-20190729/lld/ELF/Config.h
--- llvm-9.0.0-20190729/lld/ELF/Config.h.19~	2019-07-29 19:25:03.256079324 +0200
+++ llvm-9.0.0-20190729/lld/ELF/Config.h	2019-07-29 19:29:35.834477088 +0200
@@ -124,12 +124,12 @@ struct Configuration {
                   uint64_t>
       callGraphProfile;
   bool allowMultipleDefinition;
-  bool allowShlibUndefined;
+  bool allowShlibUndefined = false;
   bool androidPackDynRelocs;
   bool armHasBlx = false;
   bool armHasMovtMovw = false;
   bool armJ1J2BranchEncoding = false;
-  bool asNeeded = false;
+  bool asNeeded = true;
   bool bsymbolic;
   bool bsymbolicFunctions;
   bool callGraphProfileSort;
@@ -143,16 +143,16 @@ struct Configuration {
   bool ehFrameHdr;
   bool emitLLVM;
   bool emitRelocs;
-  bool enableNewDtags;
+  bool enableNewDtags = true;
   bool executeOnly;
   bool exportDynamic;
   bool fixCortexA53Errata843419;
   bool forceBTI;
   bool formatBinary = false;
   bool requireCET;
-  bool gcSections;
+  bool gcSections = true;
   bool gdbIndex;
-  bool gnuHash = false;
+  bool gnuHash = true;
   bool gnuUnique;
   bool hasDynamicList = false;
   bool hasDynSymTab;
@@ -189,14 +189,14 @@ struct Configuration {
   bool undefinedVersion;
   bool useAndroidRelrTags = false;
   bool warnBackrefs;
-  bool warnCommon;
+  bool warnCommon = true;
   bool warnIfuncTextrel;
   bool warnMissingEntry;
   bool warnSymbolOrdering;
   bool writeAddends;
   bool zCombreloc;
   bool zCopyreloc;
-  bool zExecstack;
+  bool zExecstack = false;
   bool zGlobal;
   bool zHazardplt;
   bool zIfuncNoplt;
@@ -214,14 +214,14 @@ struct Configuration {
   bool zRetpolineplt;
   bool zWxneeded;
   DiscardPolicy discard;
-  ICFLevel icf;
+  ICFLevel icf = ICFLevel::Safe;
   OrphanHandlingPolicy orphanHandling;
   SortSectionPolicy sortSection;
   StripPolicy strip;
   UnresolvedPolicy unresolvedSymbols;
   Target2Policy target2;
   ARMVFPArgKind armVFPArgs = ARMVFPArgKind::Default;
-  BuildIdKind buildId = BuildIdKind::None;
+  BuildIdKind buildId = BuildIdKind::Sha1;
   ELFKind ekind = ELFNoneKind;
   uint16_t defaultSymbolVersion = llvm::ELF::VER_NDX_GLOBAL;
   uint16_t emachine = llvm::ELF::EM_NONE;
@@ -232,7 +232,7 @@ struct Configuration {
   uint64_t zStackSize;
   unsigned ltoPartitions;
   unsigned ltoo;
-  unsigned optimize;
+  unsigned optimize = 1;
   unsigned thinLTOJobs;
   int32_t splitStackAdjustSize;
 
diff -up llvm-9.0.0-20190729/lld/ELF/Driver.cpp.19~ llvm-9.0.0-20190729/lld/ELF/Driver.cpp
--- llvm-9.0.0-20190729/lld/ELF/Driver.cpp.19~	2019-07-29 19:25:03.257079318 +0200
+++ llvm-9.0.0-20190729/lld/ELF/Driver.cpp	2019-07-29 19:31:32.704783589 +0200
@@ -569,10 +569,10 @@ static StringRef getDynamicLinker(opt::I
 
 static ICFLevel getICF(opt::InputArgList &args) {
   auto *arg = args.getLastArg(OPT_icf_none, OPT_icf_safe, OPT_icf_all);
-  if (!arg || arg->getOption().getID() == OPT_icf_none)
-    return ICFLevel::None;
-  if (arg->getOption().getID() == OPT_icf_safe)
+  if (!arg || arg->getOption().getID() == OPT_icf_safe)
     return ICFLevel::Safe;
+  if (arg->getOption().getID() == OPT_icf_none)
+    return ICFLevel::None;
   return ICFLevel::All;
 }
 
@@ -648,7 +648,7 @@ getBuildId(opt::InputArgList &args) {
     return {BuildIdKind::None, {}};
 
   if (arg->getOption().getID() == OPT_build_id)
-    return {BuildIdKind::Fast, {}};
+    return {BuildIdKind::Sha1, {}};
 
   StringRef s = arg->getValue();
   if (s == "fast")
@@ -831,7 +831,7 @@ static void readConfigs(opt::InputArgLis
   config->fixCortexA53Errata843419 = args.hasArg(OPT_fix_cortex_a53_843419);
   config->forceBTI = args.hasArg(OPT_force_bti);
   config->requireCET = args.hasArg(OPT_require_cet);
-  config->gcSections = args.hasFlag(OPT_gc_sections, OPT_no_gc_sections, false);
+  config->gcSections = args.hasFlag(OPT_gc_sections, OPT_no_gc_sections, true);
   config->gnuUnique = args.hasFlag(OPT_gnu_unique, OPT_no_gnu_unique, true);
   config->gdbIndex = args.hasFlag(OPT_gdb_index, OPT_no_gdb_index, false);
   config->icf = getICF(args);
@@ -912,7 +912,7 @@ static void readConfigs(opt::InputArgLis
   config->unresolvedSymbols = getUnresolvedSymbolPolicy(args);
   config->warnBackrefs =
       args.hasFlag(OPT_warn_backrefs, OPT_no_warn_backrefs, false);
-  config->warnCommon = args.hasFlag(OPT_warn_common, OPT_no_warn_common, false);
+  config->warnCommon = args.hasFlag(OPT_warn_common, OPT_no_warn_common, true);
   config->warnIfuncTextrel =
       args.hasFlag(OPT_warn_ifunc_textrel, OPT_no_warn_ifunc_textrel, false);
   config->warnSymbolOrdering =
@@ -1673,7 +1673,7 @@ template <class ELFT> void LinkerDriver:
     if (config->emachine == EM_MIPS)
       config->sysvHash = true;
     else
-      config->sysvHash = config->gnuHash = true;
+      config->sysvHash = false; config->gnuHash = true;
   }
 
   // Default output filename is "a.out" by the Unix tradition.
