diff -up llvm-project-release-10.x/lld/ELF/Driver.cpp.20~ llvm-project-release-10.x/lld/ELF/Driver.cpp
--- llvm-project-release-10.x/lld/ELF/Driver.cpp.20~	2020-02-03 16:14:10.615138009 +0100
+++ llvm-project-release-10.x/lld/ELF/Driver.cpp	2020-02-03 16:16:06.678143127 +0100
@@ -326,12 +326,16 @@ static void checkOptions() {
   if (config->relocatable) {
     if (config->shared)
       error("-r and -shared may not be used together");
-    if (config->gcSections)
-      error("-r and --gc-sections may not be used together");
+    if (config->gcSections) {
+      warn("-r and --gc-sections may not be used together, disabling --gc-sections");
+      config->gcSections = false;
+    }
     if (config->gdbIndex)
       error("-r and --gdb-index may not be used together");
-    if (config->icf != ICFLevel::None)
-      error("-r and --icf may not be used together");
+    if (config->icf != ICFLevel::None) {
+      warn("-r and --icf may not be used together, disabling --icf");
+      config->icf = ICFLevel::None;
+    }
     if (config->pie)
       error("-r and -pie may not be used together");
     if (config->exportDynamic)
