diff -up llvm-project-main/lld/ELF/Driver.cpp.8~ llvm-project-main/lld/ELF/Driver.cpp
--- llvm-project-main/lld/ELF/Driver.cpp.8~	2022-01-18 18:33:45.929360656 +0100
+++ llvm-project-main/lld/ELF/Driver.cpp	2022-01-18 18:34:59.342814459 +0100
@@ -1475,7 +1475,10 @@ static void setConfigs(opt::InputArgList
 static bool isFormatBinary(StringRef s) {
   if (s == "binary")
     return true;
-  if (s == "elf" || s == "default")
+  // OMV: Change s == "elf" to s.starts_with_insensitive("elf")
+  // for better compatibility with GNU ld (which allows for
+  // things like elf64-x86-64). Relied on by e.g. fpc
+  if (s.starts_with_insensitive("elf") || s == "default")
     return false;
   error("unknown --format value: " + s +
         " (supported formats: elf, default, binary)");
