diff -up llvm-project-llvmorg-20.1.3/lld/ELF/Driver.cpp.9~ llvm-project-llvmorg-20.1.3/lld/ELF/Driver.cpp
--- llvm-project-llvmorg-20.1.3/lld/ELF/Driver.cpp.9~	2025-04-16 02:23:49.000000000 +0200
+++ llvm-project-llvmorg-20.1.3/lld/ELF/Driver.cpp	2025-04-26 13:37:59.888740409 +0200
@@ -1924,7 +1924,10 @@ static void setConfigs(Ctx &ctx, opt::In
 static bool isFormatBinary(Ctx &ctx, StringRef s) {
   if (s == "binary")
     return true;
-  if (s == "elf" || s == "default")
+  // OMV: Change s == "elf" to s.starts_with_insensitive("elf")
+  // for better compatibility with GNU ld (which allows for
+  // things like elf64-x86-64). Relied on by e.g. fpc
+  if (s.starts_with_insensitive("elf") || s == "default")
     return false;
   ErrAlways(ctx) << "unknown --format value: " << s
                  << " (supported formats: elf, default, binary)";
