--- llvm-5.0.0.src/tools/lld/ELF/CMakeLists.txt.lldcompile~	2017-06-17 13:58:16.615078449 +0200
+++ llvm-5.0.0.src/tools/lld/ELF/CMakeLists.txt	2017-06-17 13:58:58.537104895 +0200
@@ -6,6 +6,9 @@ if(NOT LLD_BUILT_STANDALONE)
   set(tablegen_deps intrinsics_gen)
 endif()
 
+# Without this, fails to build with libstdc++ 6.2.1
+add_definitions(-std=c++11)
+
 add_lld_library(lldELF
   Arch/AArch64.cpp
   Arch/AMDGPU.cpp
