diff -up llvm-project-llvmorg-20.1.3/lld/ELF/Driver.cpp.21~ llvm-project-llvmorg-20.1.3/lld/ELF/Driver.cpp
--- llvm-project-llvmorg-20.1.3/lld/ELF/Driver.cpp.21~	2025-04-26 13:47:28.472623359 +0200
+++ llvm-project-llvmorg-20.1.3/lld/ELF/Driver.cpp	2025-04-26 13:50:37.337626610 +0200
@@ -360,9 +360,11 @@ static void initLLVM() {
 static void checkOptions(Ctx &ctx) {
   // The MIPS ABI as of 2016 does not support the GNU-style symbol lookup
   // table which is a relatively new feature.
-  if (ctx.arg.emachine == EM_MIPS && ctx.arg.gnuHash)
-    ErrAlways(ctx)
-        << "the .gnu.hash section is not compatible with the MIPS target";
+  if (ctx.arg.emachine == EM_MIPS && ctx.arg.gnuHash) {
+    Warn(ctx)
+        << "the .gnu.hash section is not compatible with the MIPS target, disabling";
+    ctx.arg.gnuHash = false;
+  }
 
   if (ctx.arg.emachine == EM_ARM) {
     if (!ctx.arg.cmseImplib) {
@@ -422,8 +424,14 @@ static void checkOptions(Ctx &ctx) {
       ErrAlways(ctx) << "-r and -shared may not be used together";
     if (ctx.arg.gdbIndex)
       ErrAlways(ctx) << "-r and --gdb-index may not be used together";
-    if (ctx.arg.icf != ICFLevel::None)
-      ErrAlways(ctx) << "-r and --icf may not be used together";
+    if (ctx.arg.gcSections) {
+      Warn(ctx) << "-r and --gc-sections may not be used together, disabling --gc-sections";
+      ctx.arg.gcSections = false;
+    }
+    if (ctx.arg.icf != ICFLevel::None) {
+      Warn(ctx) << "-r and --icf may not be used together, disabling --icf";
+      ctx.arg.icf = ICFLevel::None;
+    }
     if (ctx.arg.pie)
       ErrAlways(ctx) << "-r and -pie may not be used together";
     if (ctx.arg.exportDynamic)
