--- llvm-6.0.0.src/lld/ELF/CMakeLists.txt.lldcompile~	2017-12-08 22:42:11.494608930 +0100
+++ llvm-6.0.0.src/lld/ELF/CMakeLists.txt	2017-12-08 22:42:48.027676457 +0100
@@ -6,6 +6,9 @@ if(NOT LLD_BUILT_STANDALONE)
   set(tablegen_deps intrinsics_gen)
 endif()
 
+# Without this, fails to build with libstdc++ 6.2.1
+add_definitions(-std=c++11)
+
 add_lld_library(lldELF
   AArch64ErrataFix.cpp
   Arch/AArch64.cpp
