diff -up llvm-9.0.0-20190729/lld/ELF/Driver.cpp.22~ llvm-9.0.0-20190729/lld/ELF/Driver.cpp
--- llvm-9.0.0-20190729/lld/ELF/Driver.cpp.22~	2019-07-29 19:32:42.451368350 +0200
+++ llvm-9.0.0-20190729/lld/ELF/Driver.cpp	2019-07-29 19:34:39.745668143 +0200
@@ -320,12 +320,16 @@ static void checkOptions() {
   if (config->relocatable) {
     if (config->shared)
       error("-r and -shared may not be used together");
-    if (config->gcSections)
-      error("-r and --gc-sections may not be used together");
+    if (config->gcSections) {
+      warn("-r and --gc-sections may not be used together, disabling --gc-sections");
+      config->gcSections = false;
+    }
     if (config->gdbIndex)
       error("-r and --gdb-index may not be used together");
-    if (config->icf != ICFLevel::None)
-      error("-r and --icf may not be used together");
+    if (config->icf != ICFLevel::None) {
+      warn("-r and --icf may not be used together, disabling --icf");
+      config->icf = ICFLevel::None;
+    }
     if (config->pie)
       error("-r and -pie may not be used together");
   }
