--- llvm-project-release-12.x/lld/ELF/Driver.cpp.18~	2021-03-04 12:21:14.000000000 +0000
+++ llvm-project-release-12.x/lld/ELF/Driver.cpp	2021-03-30 11:18:46.625712292 +0000
@@ -311,8 +311,10 @@
 static void checkOptions() {
   // The MIPS ABI as of 2016 does not support the GNU-style symbol lookup
   // table which is a relatively new feature.
-  if (config->emachine == EM_MIPS && config->gnuHash)
-    error("the .gnu.hash section is not compatible with the MIPS target");
+  if (config->emachine == EM_MIPS && config->gnuHash) {
+    warn("the .gnu.hash section is not compatible with the MIPS target, disabling");
+    config->gnuHash = false;
+  }
 
   if (config->fixCortexA53Errata843419 && config->emachine != EM_AARCH64)
     error("--fix-cortex-a53-843419 is only supported on AArch64 targets");
@@ -349,8 +351,14 @@
       error("-r and -shared may not be used together");
     if (config->gdbIndex)
       error("-r and --gdb-index may not be used together");
-    if (config->icf != ICFLevel::None)
-      error("-r and --icf may not be used together");
+    if (config->gcSections) {
+      warn("-r and --gc-sections may not be used together, disabling --gc-sections");
+      config->gcSections = false;
+    }
+    if (config->icf != ICFLevel::None) {
+      warn("-r and --icf may not be used together, disabling --icf");
+      config->icf = ICFLevel::None;
+    }
     if (config->pie)
       error("-r and -pie may not be used together");
     if (config->exportDynamic)
