--- llvm-4.0.0.src/tools/lld/ELF/CMakeLists.txt.lldcompile~	2016-12-30 22:54:04.832455930 +0100
+++ llvm-4.0.0.src/tools/lld/ELF/CMakeLists.txt	2016-12-30 23:20:36.622302095 +0100
@@ -6,6 +6,9 @@ if(NOT LLD_BUILT_STANDALONE)
   set(tablegen_deps intrinsics_gen)
 endif()
 
+# Without this, fails to build with libstdc++ 6.2.1
+add_definitions(-std=c++11)
+
 add_lld_library(lldELF
   Driver.cpp
   DriverUtils.cpp
