diff -up llvm-project-17.0.6.src/lld/ELF/Driver.cpp.omv~ llvm-project-17.0.6.src/lld/ELF/Driver.cpp
--- llvm-project-17.0.6.src/lld/ELF/Driver.cpp.omv~	2024-01-02 23:40:33.278538617 +0100
+++ llvm-project-17.0.6.src/lld/ELF/Driver.cpp	2024-01-02 23:40:59.246840524 +0100
@@ -1351,7 +1351,7 @@ static void readConfigs(opt::InputArgLis
   config->trace = args.hasArg(OPT_trace);
   config->undefined = args::getStrings(args, OPT_undefined);
   config->undefinedVersion =
-      args.hasFlag(OPT_undefined_version, OPT_no_undefined_version, false);
+      args.hasFlag(OPT_undefined_version, OPT_no_undefined_version, true);
   config->unique = args.hasArg(OPT_unique);
   config->useAndroidRelrTags = args.hasFlag(
       OPT_use_android_relr_tags, OPT_no_use_android_relr_tags, false);
diff -up llvm-project-17.0.6.src/lld/ELF/Options.td.omv~ llvm-project-17.0.6.src/lld/ELF/Options.td
--- llvm-project-17.0.6.src/lld/ELF/Options.td.omv~	2024-01-02 23:39:56.506111026 +0100
+++ llvm-project-17.0.6.src/lld/ELF/Options.td	2024-01-02 23:41:13.059001099 +0100
@@ -472,7 +472,7 @@ defm unresolved_symbols:
   Eq<"unresolved-symbols", "Determine how to handle unresolved symbols">;
 
 defm undefined_version: B<"undefined-version",
-  "Allow unused version in version script (disabled by default)",
+  "Allow unused version in version script (enabled by default)",
   "Report version scripts that refer undefined symbols">;
 
 defm rsp_quoting: EEq<"rsp-quoting", "Quoting style for response files">,
