--- llvm-project-release-12.x/lld/ELF/Config.h.15~	2021-04-06 16:38:18.000000000 +0000
+++ llvm-project-release-12.x/lld/ELF/Config.h	2021-04-13 09:28:08.317981899 +0000
@@ -142,7 +142,7 @@
   bool armHasBlx = false;
   bool armHasMovtMovw = false;
   bool armJ1J2BranchEncoding = false;
-  bool asNeeded = false;
+  bool asNeeded = true;
   bool bsymbolic;
   bool bsymbolicFunctions;
   bool callGraphProfileSort;
@@ -157,7 +157,7 @@
   bool ehFrameHdr;
   bool emitLLVM;
   bool emitRelocs;
-  bool enableNewDtags;
+  bool enableNewDtags = true;
   bool executeOnly;
   bool exportDynamic;
   bool fixCortexA53Errata843419;
@@ -166,7 +166,7 @@
   bool fortranCommon;
   bool gcSections;
   bool gdbIndex;
-  bool gnuHash = false;
+  bool gnuHash = true;
   bool gnuUnique;
   bool hasDynSymTab;
   bool ignoreDataAddressEquality;
@@ -214,7 +214,7 @@
   bool useAndroidRelrTags = false;
   bool warnBackrefs;
   std::vector<llvm::GlobPattern> warnBackrefsExclude;
-  bool warnCommon;
+  bool warnCommon = true;
   bool warnIfuncTextrel;
   bool warnMissingEntry;
   bool warnSymbolOrdering;
@@ -252,7 +252,7 @@
   UnresolvedPolicy unresolvedSymbolsInShlib;
   Target2Policy target2;
   ARMVFPArgKind armVFPArgs = ARMVFPArgKind::Default;
-  BuildIdKind buildId = BuildIdKind::None;
+  BuildIdKind buildId = BuildIdKind::Sha1;
   SeparateSegmentKind zSeparate;
   ELFKind ekind = ELFNoneKind;
   uint16_t emachine = llvm::ELF::EM_NONE;
@@ -263,7 +263,7 @@
   uint64_t zStackSize;
   unsigned ltoPartitions;
   unsigned ltoo;
-  unsigned optimize;
+  unsigned optimize = 1;
   StringRef thinLTOJobs;
   unsigned timeTraceGranularity;
   int32_t splitStackAdjustSize;
--- llvm-project-release-12.x/lld/ELF/Driver.cpp.15~	2021-04-06 16:38:18.000000000 +0000
+++ llvm-project-release-12.x/lld/ELF/Driver.cpp	2021-04-13 09:40:15.567246860 +0000
@@ -770,7 +770,7 @@
     return {BuildIdKind::None, {}};
 
   if (arg->getOption().getID() == OPT_build_id)
-    return {BuildIdKind::Fast, {}};
+    return {BuildIdKind::Sha1, {}};
 
   StringRef s = arg->getValue();
   if (s == "fast")
@@ -1092,7 +1092,7 @@
       OPT_use_android_relr_tags, OPT_no_use_android_relr_tags, false);
   config->warnBackrefs =
       args.hasFlag(OPT_warn_backrefs, OPT_no_warn_backrefs, false);
-  config->warnCommon = args.hasFlag(OPT_warn_common, OPT_no_warn_common, false);
+  config->warnCommon = args.hasFlag(OPT_warn_common, OPT_no_warn_common, true);
   config->warnIfuncTextrel =
       args.hasFlag(OPT_warn_ifunc_textrel, OPT_no_warn_ifunc_textrel, false);
   config->warnSymbolOrdering =
@@ -2041,7 +2041,7 @@
     if (config->emachine == EM_MIPS)
       config->sysvHash = true;
     else
-      config->sysvHash = config->gnuHash = true;
+      {config->sysvHash = false; config->gnuHash = true;}
   }
 
   // Default output filename is "a.out" by the Unix tradition.
