--- llvm-project-llvmorg-23.1.0-rc1/lldb/source/API/CMakeLists.txt
+++ llvm-project-llvmorg-23.1.0-rc1/lldb/source/API/CMakeLists.txt
@@ -145,6 +145,8 @@ add_lldb_library(liblldb SHARED ${option
 if (NOT LLDB_ENABLE_DYNAMIC_SCRIPTINTERPRETERS)
   if (LLDB_ENABLE_PYTHON)
     add_python_wrapper(liblldb)
+    # Workaround for swig issue
+    add_definitions(-Wno-error=format-security)
   endif()
 
   if(LLDB_ENABLE_LUA)
