--- llvm-4.0.0.src/tools/lldb/scripts/CMakeLists.txt.lldb~	2017-01-20 18:24:03.812704736 +0100
+++ llvm-4.0.0.src/tools/lldb/scripts/CMakeLists.txt	2017-01-20 18:40:12.006753964 +0100
@@ -51,6 +51,9 @@ add_custom_command(
 set_source_files_properties(${LLDB_WRAP_PYTHON} PROPERTIES GENERATED 1)
 set_source_files_properties(${CMAKE_CURRENT_BINARY_DIR}/lldb.py PROPERTIES GENERATED 1)
 
+# Needed because of the way swig wraps SetErrorStringWithFormat
+add_definitions(-Wno-error=format-security)
+
 add_custom_target(swig_wrapper ALL DEPENDS ${LLDB_WRAP_PYTHON})
 
 # Install the LLDB python module
--- llvm-4.0.0.src/tools/lldb/source/API/CMakeLists.txt.lldb~	2016-12-15 19:18:47.000000000 +0100
+++ llvm-4.0.0.src/tools/lldb/source/API/CMakeLists.txt	2017-01-20 18:24:03.813704740 +0100
@@ -90,6 +90,9 @@ if (LLVM_ENABLE_WERROR)
   endif()
 endif()
 
+# Needed because of the way swig wraps SetErrorStringWithFormat
+add_definitions(-Wno-error=format-security)
+
 # This should not be part of LLDBDependencies.cmake, because we don't
 # want every single library taking a dependency on the script interpreters.
 target_link_libraries(liblldb PRIVATE
