diff -up llvm-project-release-17.x/lldb/CMakeLists.txt.39~ llvm-project-release-17.x/lldb/CMakeLists.txt
--- llvm-project-release-17.x/lldb/CMakeLists.txt.39~	2023-07-28 13:26:44.435790234 +0200
+++ llvm-project-release-17.x/lldb/CMakeLists.txt	2023-07-28 13:27:18.718058950 +0200
@@ -21,6 +21,12 @@ if (CMAKE_SOURCE_DIR STREQUAL CMAKE_CURR
   set(LLVM_INCLUDE_TESTS ON CACHE INTERNAL "")
 endif()
 
+if("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "riscv64")
+  set(atomic_libs "atomic")
+else()
+  set(atomic_libs)
+endif()
+
 # Must go below project(..)
 include(GNUInstallDirs)
 
diff -up llvm-project-release-17.x/lldb/source/API/CMakeLists.txt.39~ llvm-project-release-17.x/lldb/source/API/CMakeLists.txt
--- llvm-project-release-17.x/lldb/source/API/CMakeLists.txt.39~	2023-07-28 13:26:43.323781550 +0200
+++ llvm-project-release-17.x/lldb/source/API/CMakeLists.txt	2023-07-28 13:26:44.436790242 +0200
@@ -109,6 +109,7 @@ add_lldb_library(liblldb SHARED ${option
     lldbUtility
     lldbVersion
     ${LLDB_ALL_PLUGINS}
+    ${atomic_libs}
   LINK_COMPONENTS
     Support
 
diff -up llvm-project-release-17.x/lldb/tools/lldb-server/CMakeLists.txt.39~ llvm-project-release-17.x/lldb/tools/lldb-server/CMakeLists.txt
--- llvm-project-release-17.x/lldb/tools/lldb-server/CMakeLists.txt.39~	2023-07-28 09:10:13.000000000 +0200
+++ llvm-project-release-17.x/lldb/tools/lldb-server/CMakeLists.txt	2023-07-28 13:26:44.436790242 +0200
@@ -56,6 +56,7 @@ add_lldb_tool(lldb-server
       lldbPluginInstructionMIPS64
       lldbPluginInstructionRISCV
       ${LLDB_SYSTEM_LIBS}
+      ${atomic_libs}
 
     LINK_COMPONENTS
       Option
