diff -up llvm-project-release-12.x/lldb/source/API/CMakeLists.txt.10~ llvm-project-release-12.x/lldb/source/API/CMakeLists.txt
--- llvm-project-release-12.x/lldb/source/API/CMakeLists.txt.10~	2021-02-08 19:19:45.320197985 +0100
+++ llvm-project-release-12.x/lldb/source/API/CMakeLists.txt	2021-02-08 19:21:03.304516547 +0100
@@ -129,6 +129,8 @@ endif()
 
 if(LLDB_ENABLE_PYTHON)
   add_dependencies(liblldb swig_wrapper_python)
+  # Workaround for swig issue
+  add_definitions(-Wno-error=format-security)
 
   if (MSVC)
     set_property(SOURCE ${lldb_python_wrapper} APPEND_STRING PROPERTY COMPILE_FLAGS " /W0")
