--- llvm-3.5/tools/clang/lib/Driver/Tools.cpp.omv~	2014-05-26 20:50:58.405041023 +0200
+++ llvm-3.5/tools/clang/lib/Driver/Tools.cpp	2014-05-26 20:54:00.388516198 +0200
@@ -1391,7 +1391,11 @@ static void AddGoldPlugin(const ToolChai
   // as gold requires -plugin to come before any -plugin-opt that -Wl might
   // forward.
   CmdArgs.push_back("-plugin");
+#if defined(__x86_64__) || defined(__aarch64__)
+  std::string Plugin = ToolChain.getDriver().Dir + "/../lib64/LLVMgold.so";
+#else
   std::string Plugin = ToolChain.getDriver().Dir + "/../lib/LLVMgold.so";
+#endif
   CmdArgs.push_back(Args.MakeArgString(Plugin));
 
   // Try to pass driver level flags relevant to LTO code generation down to
