diff -up llvm-project-15.0.4.src/lldb/bindings/interface/SBValueList.i.omv~ llvm-project-15.0.4.src/lldb/bindings/interface/SBValueList.i
--- llvm-project-15.0.4.src/lldb/bindings/interface/SBValueList.i.omv~	2022-11-03 01:44:45.398463661 +0100
+++ llvm-project-15.0.4.src/lldb/bindings/interface/SBValueList.i	2022-11-03 01:44:52.383516610 +0100
@@ -103,7 +103,6 @@ public:
     GetFirstValueByName (const char* name) const;
 
     %extend {
-       %nothreadallow;
        std::string lldb::SBValueList::__str__ (){
            lldb::SBStream description;
            const size_t n = $self->GetSize();
@@ -122,7 +121,6 @@ public:
                --desc_len;
            return std::string(desc, desc_len);
        }
-       %clearnothreadallow;
     }
 
 #ifdef SWIGPYTHON
