diff -up llvm-project-release-17.x/llvm/include/llvm/Analysis/TargetLibraryInfo.h.18~ llvm-project-release-17.x/llvm/include/llvm/Analysis/TargetLibraryInfo.h
--- llvm-project-release-17.x/llvm/include/llvm/Analysis/TargetLibraryInfo.h.18~	2023-07-28 13:09:55.724858577 +0200
+++ llvm-project-release-17.x/llvm/include/llvm/Analysis/TargetLibraryInfo.h	2023-07-28 13:10:40.558284333 +0200
@@ -17,6 +17,17 @@
 #include "llvm/TargetParser/Triple.h"
 #include <optional>
 
+#ifndef __GLIBC__ // for musl
+#undef fopen64
+#undef fseeko64
+#undef fstat64
+#undef fstatvfs64
+#undef ftello64
+#undef lstat64
+#undef stat64
+#undef tmpfile64
+#endif
+
 namespace llvm {
 
 template <typename T> class ArrayRef;
diff -up llvm-project-release-17.x/llvm/lib/Support/Unix/Signals.inc.18~ llvm-project-release-17.x/llvm/lib/Support/Unix/Signals.inc
--- llvm-project-release-17.x/llvm/lib/Support/Unix/Signals.inc.18~	2023-07-28 13:09:55.724858577 +0200
+++ llvm-project-release-17.x/llvm/lib/Support/Unix/Signals.inc	2023-07-28 13:11:18.624645900 +0200
@@ -465,7 +465,7 @@ void llvm::sys::AddSignalHandler(sys::Si
 }
 
 #if ENABLE_BACKTRACES && defined(HAVE_BACKTRACE) && HAVE_LINK_H &&             \
-    (defined(__linux__) || defined(__FreeBSD__) ||                             \
+    ((defined(__linux__) && defined(__GLIBC__)) || defined(__FreeBSD__) ||                             \
      defined(__FreeBSD_kernel__) || defined(__NetBSD__))
 struct DlIteratePhdrData {
   void **StackTrace;
