diff -up llvm-project-release-12.x/clang/lib/Basic/Targets/OSTargets.h.27~ llvm-project-release-12.x/clang/lib/Basic/Targets/OSTargets.h
--- llvm-project-release-12.x/clang/lib/Basic/Targets/OSTargets.h.27~	2021-02-08 19:45:41.223374831 +0100
+++ llvm-project-release-12.x/clang/lib/Basic/Targets/OSTargets.h	2021-02-08 19:47:06.882909389 +0100
@@ -256,10 +256,13 @@ public:
     case llvm::Triple::ppcle:
     case llvm::Triple::ppc64:
     case llvm::Triple::ppc64le:
+    case llvm::Triple::aarch64:
+    case llvm::Triple::aarch64_be:
       this->MCountName = "_mcount";
       break;
     case llvm::Triple::arm:
-      this->MCountName = "__mcount";
+    case llvm::Triple::armeb:
+      this->MCountName = "__gnu_mcount_nc";
       break;
     }
   }
