--- llvm-3.9.0.src/tools/clang/lib/Basic/Targets.cpp.mcount~	2016-06-18 13:34:35.412141933 +0200
+++ llvm-3.9.0.src/tools/clang/lib/Basic/Targets.cpp	2016-06-18 13:35:47.395296695 +0200
@@ -358,10 +358,13 @@ public:
     case llvm::Triple::ppc:
     case llvm::Triple::ppc64:
     case llvm::Triple::ppc64le:
+    case llvm::Triple::aarch64:
+    case llvm::Triple::aarch64_be:
       this->MCountName = "_mcount";
       break;
     case llvm::Triple::arm:
-      this->MCountName = "__mcount";
+    case llvm::Triple::armeb:
+      this->MCountName = "__gnu_mcount_nc";
       break;
     }
   }
