--- llvm-4.0.0.src/tools/lldb/cmake/modules/AddLLDB.cmake.lldbstatic~	2016-12-30 22:12:17.924867093 +0100
+++ llvm-4.0.0.src/tools/lldb/cmake/modules/AddLLDB.cmake	2016-12-30 22:52:33.551323736 +0100
@@ -34,19 +34,7 @@ function(add_lldb_library name)
       ../../include/lldb${dir}/*.h)
     set(srcs ${srcs} ${headers})
   endif()
-  if (PARAM_MODULE)
-    set(libkind MODULE)
-  elseif (PARAM_SHARED)
-    set(libkind SHARED)
-  elseif (PARAM_OBJECT)
-    set(libkind OBJECT)
-  else ()
-    # PARAM_STATIC or library type unspecified. BUILD_SHARED_LIBS
-    # does not control the kind of libraries created for LLDB,
-    # only whether or not they link to shared/static LLVM/Clang
-    # libraries.
-    set(libkind STATIC)
-  endif()
+  set(libkind STATIC)
 
   #PIC not needed on Win
   if (NOT WIN32)
