diff -up llvm-project-release-16.x/llvm/lib/Support/Unix/Program.inc.26~ llvm-project-release-16.x/llvm/lib/Support/Unix/Program.inc
--- llvm-project-release-16.x/llvm/lib/Support/Unix/Program.inc.26~	2023-01-30 17:54:55.234917668 +0100
+++ llvm-project-release-16.x/llvm/lib/Support/Unix/Program.inc	2023-01-30 17:56:44.017525611 +0100
@@ -261,7 +261,7 @@ static bool Execute(ProcessInfo &PI, Str
       posix_spawn_file_actions_destroy(FileActions);
 
     if (Err)
-      return !MakeErrMsg(ErrMsg, "posix_spawn failed", Err);
+      return !MakeErrMsg(ErrMsg, "posix_spawn " + Program.str() + " failed", Err);
 
     PI.Pid = PID;
     PI.Process = PID;
