diff -up llvm-7.0.0.src/projects/compiler-rt/lib/sanitizer_common/sanitizer_platform_limits_linux.cc.crtBuildfix~ llvm-7.0.0.src/projects/compiler-rt/lib/sanitizer_common/sanitizer_platform_limits_linux.cc
--- llvm-7.0.0.src/projects/compiler-rt/lib/sanitizer_common/sanitizer_platform_limits_linux.cc.crtBuildfix~	2018-07-07 13:24:45.140762506 +0200
+++ llvm-7.0.0.src/projects/compiler-rt/lib/sanitizer_common/sanitizer_platform_limits_linux.cc	2018-07-07 13:28:50.747771665 +0200
@@ -49,7 +49,11 @@
 #undef off_t
 #endif
 
+#include <time.h>
+#define _STRUCT_TIMESPEC
+#define itimerspec kernel_itimerspec
 #include <linux/aio_abi.h>
+#undef itimerspec
 
 #if !SANITIZER_ANDROID
 #include <sys/statfs.h>
