diff -up llvm-project-master/llvm/include/llvm/Analysis/TargetLibraryInfo.h.15~ llvm-project-master/llvm/include/llvm/Analysis/TargetLibraryInfo.h
--- llvm-project-master/llvm/include/llvm/Analysis/TargetLibraryInfo.h.15~	2020-06-21 19:57:56.204244940 +0200
+++ llvm-project-master/llvm/include/llvm/Analysis/TargetLibraryInfo.h	2020-06-21 19:58:43.362816675 +0200
@@ -18,6 +18,17 @@
 #include "llvm/IR/PassManager.h"
 #include "llvm/Pass.h"
 
+#ifndef __GLIBC__ // for musl
+#undef fopen64
+#undef fseeko64
+#undef fstat64
+#undef fstatvfs64
+#undef ftello64
+#undef lstat64
+#undef stat64
+#undef tmpfile64
+#endif
+
 namespace llvm {
 template <typename T> class ArrayRef;
 class Triple;
diff -up llvm-project-master/llvm/lib/Support/Unix/Signals.inc.15~ llvm-project-master/llvm/lib/Support/Unix/Signals.inc
--- llvm-project-master/llvm/lib/Support/Unix/Signals.inc.15~	2020-06-21 19:00:42.000000000 +0200
+++ llvm-project-master/llvm/lib/Support/Unix/Signals.inc	2020-06-21 19:57:56.204244940 +0200
@@ -460,7 +460,7 @@ void llvm::sys::AddSignalHandler(sys::Si
   RegisterHandlers();
 }
 
-#if defined(HAVE_BACKTRACE) && ENABLE_BACKTRACES && HAVE_LINK_H &&    \
+#if defined(HAVE_BACKTRACE) && defined(__GLIBC__) && ENABLE_BACKTRACES && HAVE_LINK_H &&    \
     (defined(__linux__) || defined(__FreeBSD__) ||                             \
      defined(__FreeBSD_kernel__) || defined(__NetBSD__))
 struct DlIteratePhdrData {
