--- llvm-3.7.0.src/lib/Support/Triple.cpp.triplet~	2015-05-20 19:26:02.907626100 +0200
+++ llvm-3.7.0.src/lib/Support/Triple.cpp	2015-05-20 19:28:03.199018763 +0200
@@ -286,6 +286,7 @@ static Triple::ArchType parseARMArch(Str
     .Cases("v6m", "v6sm", arch)
     .Cases("v6t2", "v6z", "v6zk", arch)
     .Cases("v7", "v7a", "v7em", "v7hl", "v7l", arch)
+    .Cases("v7hl", "v7nhl", arch)
     .Cases("v7m", "v7r", "v7s", arch)
     .Cases("v8", "v8a", arch)
     .Cases("v8.1", "v8.1a", arch)
