diff -up llvm-project-main/openmp/runtime/src/CMakeLists.txt.23~ llvm-project-main/openmp/runtime/src/CMakeLists.txt
--- llvm-project-main/openmp/runtime/src/CMakeLists.txt.23~	2022-01-18 18:42:17.410313634 +0100
+++ llvm-project-main/openmp/runtime/src/CMakeLists.txt	2022-01-18 18:43:27.943693120 +0100
@@ -158,6 +158,14 @@ else()
   set(LIBOMP_LINKER_LANGUAGE CXX)
 endif()
 
+if (UNIX)
+  if (NOT APPLE)
+    set_target_properties(omp PROPERTIES
+                              SOVERSION 1
+			      VERSION 1.${LLVM_VERSION_MAJOR}.${LLVM_VERSION_MINOR}.${LLVM_VERSION_PATCH}${LLVM_VERSION_SUFFIX})
+  endif()
+endif()
+
 if(OPENMP_MSVC_NAME_SCHEME)
   if(uppercase_CMAKE_BUILD_TYPE STREQUAL "DEBUG")
     set(LIBOMP_PDB_NAME ${LIBOMP_DEFAULT_LIB_NAME}${MSVC_TOOLS_VERSION}d.${LIBOMP_ARCH})
