diff -up llvm-project-release-17.x/llvm/tools/gold/CMakeLists.txt.16~ llvm-project-release-17.x/llvm/tools/gold/CMakeLists.txt
--- llvm-project-release-17.x/llvm/tools/gold/CMakeLists.txt.16~	2023-07-28 09:10:13.000000000 +0200
+++ llvm-project-release-17.x/llvm/tools/gold/CMakeLists.txt	2023-07-28 12:59:18.228821727 +0200
@@ -1,7 +1,9 @@
 set(LLVM_EXPORTED_SYMBOL_FILE ${CMAKE_CURRENT_SOURCE_DIR}/gold.exports)
 
 if( LLVM_ENABLE_PIC AND LLVM_BINUTILS_INCDIR )
-  include_directories( ${LLVM_BINUTILS_INCDIR} )
+  # Upstream does include_directories(${LLVM_BINUTILS_INCDIR}), but
+  # -I/usr/include is evil for various reasons (esp. when crosscompiling)
+  add_definitions( "-DBINUTILS_HEADER=\"${LLVM_BINUTILS_INCDIR}/plugin-api.h\"" )
 
   set(LLVM_LINK_COMPONENTS
      ${LLVM_TARGETS_TO_BUILD}
diff -up llvm-project-release-17.x/llvm/tools/gold/gold-plugin.cpp.16~ llvm-project-release-17.x/llvm/tools/gold/gold-plugin.cpp
--- llvm-project-release-17.x/llvm/tools/gold/gold-plugin.cpp.16~	2023-07-28 12:59:18.228821727 +0200
+++ llvm-project-release-17.x/llvm/tools/gold/gold-plugin.cpp	2023-07-28 13:00:18.703392369 +0200
@@ -35,7 +35,7 @@
 #include "llvm/TargetParser/Host.h"
 #include <list>
 #include <map>
-#include <plugin-api.h>
+#include BINUTILS_HEADER
 #include <string>
 #include <system_error>
 #include <utility>
