--- llvm-project-llvmorg-23.1.0-rc1/libc/src/__support/OSUtil/linux/syscall_wrappers/fcntl.h
+++ llvm-project-llvmorg-23.1.0-rc1/libc/src/__support/OSUtil/linux/syscall_wrappers/fcntl.h
@@ -30,7 +30,9 @@ namespace LIBC_NAMESPACE_DECL {
 namespace linux_syscalls {
 
 LIBC_INLINE ErrorOr<int> fcntl(int fd, int cmd, void *arg = nullptr) {
-#ifdef SYS_fcntl
+#if defined(__musl__)
+  constexpr auto FCNTL_SYSCALL_ID = SYS_fcntl64;
+#elif defined(SYS_fcntl)
   constexpr auto FCNTL_SYSCALL_ID = SYS_fcntl;
 #elif defined(SYS_fcntl64)
   constexpr auto FCNTL_SYSCALL_ID = SYS_fcntl64;
@@ -38,6 +40,7 @@ LIBC_INLINE ErrorOr<int> fcntl(int fd, i
 #error "fcntl and fcntl64 syscalls not available."
 #endif
 
+#if !defined(__musl__) // The conversion shim isn't needed for musl
   switch (cmd) {
   case F_OFD_SETLKW: {
     struct flock *flk = reinterpret_cast<struct flock *>(arg);
@@ -108,6 +111,7 @@ LIBC_INLINE ErrorOr<int> fcntl(int fd, i
   }
 #endif
   }
+#endif
 
   // Plain passthrough for all other commands. When only SYS_fcntl64 is
   // available, F_GETLK/F_SETLK/F_SETLKW have been rewritten to their 64-bit
