--- llvm-project-llvmorg-23.1.0-rc1/libc/cmake/modules/LLVMLibCArchitectures.cmake
+++ llvm-project-llvmorg-23.1.0-rc1/libc/cmake/modules/LLVMLibCArchitectures.cmake
@@ -75,7 +75,7 @@ function(get_arch_and_system_from_triple
 
   # Setting OS name for GPU architectures.
   list(GET triple_comps 2 gpu_target_sys)
-  if(gpu_target_sys MATCHES "^amdhsa" OR gpu_target_sys MATCHES "^cuda" OR target_arch MATCHES "^spirv")
+  if(gpu_target_sys MATCHES "^amdhsa" OR gpu_target_sys MATCHES "^cuda" OR gpu_target_sys MATCHES "^nvcl" OR target_arch MATCHES "^spirv")
     set(target_sys "gpu")
   endif()
 
