--- llvm-project-llvmorg-23.1.0-rc1/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp
+++ llvm-project-llvmorg-23.1.0-rc1/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp
@@ -1718,14 +1718,14 @@ SDValue DAGCombiner::PromoteIntShiftOp(S
     if (!N0.getNode())
       return SDValue();
 
+    if (Replace)
+      ReplaceLoadWithPromotedLoad(Op.getOperand(0).getNode(), N0.getNode());
+
     SDLoc DL(Op);
     SDValue N1 = Op.getOperand(1);
     SDValue RV = DAG.getNode(ISD::TRUNCATE, DL, VT,
                              DAG.getNode(Opc, DL, PVT, N0, N1), TruncFlags);
 
-    if (Replace)
-      ReplaceLoadWithPromotedLoad(Op.getOperand(0).getNode(), N0.getNode());
-
     // Deal with Op being deleted.
     if (Op && Op.getOpcode() != ISD::DELETED_NODE)
       return RV;
