--- llvm-3.6.0.src/lib/Support/Triple.cpp~	2015-02-12 18:07:52.000000000 +0100
+++ llvm-3.6.0.src/lib/Support/Triple.cpp	2015-02-12 18:07:53.797779091 +0100
@@ -273,6 +273,7 @@ static Triple::ArchType parseARMArch(Str
     .Cases("v6", "v6j", "v6k", "v6m", arch)
     .Cases("v6t2", "v6z", "v6zk", arch)
     .Cases("v7", "v7a", "v7em", "v7l", arch)
+    .Cases("v7hl", "v7hnl", arch)
     .Cases("v7m", "v7r", "v7s", arch)
     .Cases("v8", "v8a", arch)
     .Default(Triple::UnknownArch);
