diff -up llvm-project-release-13.x/clang/lib/Basic/Targets/OSTargets.h.24~ llvm-project-release-13.x/clang/lib/Basic/Targets/OSTargets.h
--- llvm-project-release-13.x/clang/lib/Basic/Targets/OSTargets.h.24~	2021-08-02 22:30:34.849316363 +0200
+++ llvm-project-release-13.x/clang/lib/Basic/Targets/OSTargets.h	2021-08-02 22:32:11.189822771 +0200
@@ -246,11 +246,16 @@ public:
     case llvm::Triple::ppcle:
     case llvm::Triple::ppc64:
     case llvm::Triple::ppc64le:
+    case llvm::Triple::aarch64:
+    case llvm::Triple::aarch64_be:
       this->MCountName = "_mcount";
       break;
     case llvm::Triple::arm:
       this->MCountName = "__mcount";
       break;
+    case llvm::Triple::armeb:
+      this->MCountName = "__gnu_mcount_nc";
+      break;
     case llvm::Triple::riscv32:
     case llvm::Triple::riscv64:
       break;
