--- llvm-project-llvmorg-23.1.0-rc1/llvm/include/llvm/ADT/ArrayRef.h
+++ llvm-project-llvmorg-23.1.0-rc1/llvm/include/llvm/ADT/ArrayRef.h
@@ -99,7 +99,7 @@ public:
       : Data(Arr), Length(N) {}
 
   /// Construct an ArrayRef from a std::initializer_list.
-#if LLVM_GNUC_PREREQ(9, 0, 0)
+#if LLVM_GNUC_PREREQ(9, 0, 0) && !defined(__clang__)
 // Disable gcc's warning in this constructor as it generates an enormous amount
 // of messages. Anyone using ArrayRef should already be aware of the fact that
 // it does not do lifetime extension.
@@ -110,7 +110,7 @@ public:
       std::initializer_list<T> Vec LLVM_LIFETIME_BOUND)
       : Data(Vec.begin() == Vec.end() ? (T *)nullptr : Vec.begin()),
         Length(Vec.size()) {}
-#if LLVM_GNUC_PREREQ(9, 0, 0)
+#if LLVM_GNUC_PREREQ(9, 0, 0) && !defined(__clang__)
 #pragma GCC diagnostic pop
 #endif
 
