--- llvm-4.0.0.src/tools/clang/lib/Basic/FileManager.cpp.compile~	2016-09-08 19:01:15.032635468 +0200
+++ llvm-4.0.0.src/tools/clang/lib/Basic/FileManager.cpp	2016-09-08 19:07:19.956166652 +0200
@@ -26,6 +26,9 @@
 #include "llvm/Support/MemoryBuffer.h"
 #include "llvm/Support/Path.h"
 #include "llvm/Support/raw_ostream.h"
+#ifdef __linux__
+#include <linux/limits.h> // for PATH_MAX
+#endif
 #include <algorithm>
 #include <cassert>
 #include <climits>
