diff -up llvm-project-release-16.x/clang/lib/Driver/ToolChains/Linux.cpp.30~ llvm-project-release-16.x/clang/lib/Driver/ToolChains/Linux.cpp
--- llvm-project-release-16.x/clang/lib/Driver/ToolChains/Linux.cpp.30~	2023-01-30 17:57:19.390887117 +0100
+++ llvm-project-release-16.x/clang/lib/Driver/ToolChains/Linux.cpp	2023-01-30 18:12:33.024966634 +0100
@@ -249,7 +249,7 @@ Linux::Linux(const Driver &D, const llvm
   }
 
 #ifdef ENABLE_LINKER_BUILD_ID
-  ExtraOpts.push_back("--build-id");
+  ExtraOpts.push_back("--build-id=sha1");
 #endif
 
   // The selection of paths to try here is designed to match the patterns which
diff -up llvm-project-release-16.x/lld/ELF/Driver.cpp.30~ llvm-project-release-16.x/lld/ELF/Driver.cpp
--- llvm-project-release-16.x/lld/ELF/Driver.cpp.30~	2023-01-30 17:57:19.266885869 +0100
+++ llvm-project-release-16.x/lld/ELF/Driver.cpp	2023-01-30 17:57:19.391887127 +0100
@@ -842,7 +842,7 @@ static std::pair<BuildIdKind, SmallVecto
 getBuildId(opt::InputArgList &args) {
   auto *arg = args.getLastArg(OPT_build_id);
   if (!arg)
-    return {BuildIdKind::None, {}};
+    return {BuildIdKind::Sha1, {}};
 
   StringRef s = arg->getValue();
   if (s == "fast")
