diff -up llvm-7.0.0.src/llvm/tools/gold/gold-plugin.cpp.fixOptlevel~ llvm-7.0.0.src/tools/gold/gold-plugin.cpp
--- llvm-7.0.0.src/llvm/tools/gold/gold-plugin.cpp.fixOptlevel~	2018-04-20 12:18:36.000000000 +0200
+++ llvm-7.0.0.src/llvm/tools/gold/gold-plugin.cpp	2018-04-25 19:11:48.153310176 +0200
@@ -254,9 +254,12 @@ namespace options {
     } else if (opt.startswith("cache-policy=")) {
       cache_policy = opt.substr(strlen("cache-policy="));
     } else if (opt.size() == 2 && opt[0] == 'O') {
-      if (opt[1] < '0' || opt[1] > '3')
-        message(LDPL_FATAL, "Optimization level must be between 0 and 3");
-      OptLevel = opt[1] - '0';
+      if (opt[1] < '0')
+        OptLevel = 0;
+      else if(opt[1] > '3')
+        OptLevel = 3;
+      else
+        OptLevel = opt[1] - '0';
     } else if (opt.startswith("jobs=")) {
       if (StringRef(opt_ + 5).getAsInteger(10, Parallelism))
         message(LDPL_FATAL, "Invalid parallelism level: %s", opt_ + 5);
diff -up llvm-7.0.0.src/llvm/tools/llvm-lto/llvm-lto.cpp.fixOptlevel~ llvm-7.0.0.src/tools/llvm-lto/llvm-lto.cpp
--- llvm-7.0.0.src/llvm/tools/llvm-lto/llvm-lto.cpp.fixOptlevel~	2018-04-25 19:11:48.153310176 +0200
+++ llvm-7.0.0.src/llvm/tools/llvm-lto/llvm-lto.cpp	2018-04-25 19:13:22.568158313 +0200
@@ -800,8 +800,10 @@ int main(int argc, char **argv) {
   InitLLVM X(argc, argv);
   cl::ParseCommandLineOptions(argc, argv, "llvm LTO linker\n");
 
-  if (OptLevel < '0' || OptLevel > '3')
-    error("optimization level must be between 0 and 3");
+  if (OptLevel < '0')
+    OptLevel = '0';
+  else if (OptLevel > '3')
+    OptLevel = '3';
 
   // Initialize the configured targets.
   InitializeAllTargets();
diff -up llvm-7.0.0.src/tools/lto/lto.cpp.fixOptlevel~ llvm-7.0.0.src/tools/lto/lto.cpp
--- llvm-7.0.0.src/llvm/tools/lto/lto.cpp.fixOptlevel~	2018-04-11 20:49:37.000000000 +0200
+++ llvm-7.0.0.src/llvm/tools/lto/lto.cpp	2018-04-25 19:11:48.153310176 +0200
@@ -164,9 +164,12 @@ static void lto_add_attrs(lto_code_gen_t
     CG->setAttr(attrs);
   }
 
-  if (OptLevel < '0' || OptLevel > '3')
-    report_fatal_error("Optimization level must be between 0 and 3");
-  CG->setOptLevel(OptLevel - '0');
+  if (OptLevel < '0')
+    CG->setOptLevel(0);
+  else if (OptLevel > '3')
+    CG->setOptLevel(3);
+  else
+    CG->setOptLevel(OptLevel - '0');
   CG->setFreestanding(EnableFreestanding);
 }
 
