--- llvm-4.0.0.src/include/llvm/Analysis/TargetLibraryInfo.h.musl1~	2017-01-09 17:27:56.000000000 +0100
+++ llvm-4.0.0.src/include/llvm/Analysis/TargetLibraryInfo.h	2017-01-20 18:41:38.680243199 +0100
@@ -18,6 +18,17 @@
 #include "llvm/IR/PassManager.h"
 #include "llvm/Pass.h"
 
+#ifndef __GLIBC__ // for musl
+#undef fopen64
+#undef fseeko64
+#undef fstat64
+#undef fstatvfs64
+#undef ftello64
+#undef lstat64
+#undef stat64
+#undef tmpfile64
+#endif
+
 namespace llvm {
 template <typename T> class ArrayRef;
 
--- llvm-4.0.0.src/lib/Analysis/TargetLibraryInfo.cpp.musl1~	2016-12-16 00:45:11.000000000 +0100
+++ llvm-4.0.0.src/lib/Analysis/TargetLibraryInfo.cpp	2017-01-20 18:41:38.681243205 +0100
@@ -394,14 +394,15 @@ static void initialize(TargetLibraryInfo
   }
 
   // The following functions are available on at least Linux:
-  if (!T.isOSLinux()) {
+  if (!T.isOSLinux())
+    TLI.setUnavailable(LibFunc::memalign);
+  if (1 /*!T.isGlibc()*/) {
     TLI.setUnavailable(LibFunc::dunder_strdup);
     TLI.setUnavailable(LibFunc::dunder_strtok_r);
     TLI.setUnavailable(LibFunc::dunder_isoc99_scanf);
     TLI.setUnavailable(LibFunc::dunder_isoc99_sscanf);
     TLI.setUnavailable(LibFunc::under_IO_getc);
     TLI.setUnavailable(LibFunc::under_IO_putc);
-    TLI.setUnavailable(LibFunc::memalign);
     TLI.setUnavailable(LibFunc::fopen64);
     TLI.setUnavailable(LibFunc::fseeko64);
     TLI.setUnavailable(LibFunc::fstat64);
--- llvm-4.0.0.src/lib/Support/DynamicLibrary.cpp.musl1~	2016-11-30 16:34:29.000000000 +0100
+++ llvm-4.0.0.src/lib/Support/DynamicLibrary.cpp	2017-01-20 18:41:38.682243211 +0100
@@ -138,7 +138,7 @@ void* DynamicLibrary::SearchForAddressOf
 
 // This macro returns the address of a well-known, explicit symbol
 #define EXPLICIT_SYMBOL(SYM) \
-   if (!strcmp(symbolName, #SYM)) return &SYM
+   if (!strcmp(symbolName, #SYM)) return (void *) &SYM
 
 // On linux we have a weird situation. The stderr/out/in symbols are both
 // macros and global variables because of standards requirements. So, we
--- llvm-4.0.0.src/lib/Support/Unix/Signals.inc.musl1~	2017-01-06 03:26:33.000000000 +0100
+++ llvm-4.0.0.src/lib/Support/Unix/Signals.inc	2017-01-20 18:41:38.682243211 +0100
@@ -298,7 +298,7 @@ void llvm::sys::AddSignalHandler(void (*
   RegisterHandlers();
 }
 
-#if defined(HAVE_BACKTRACE) && ENABLE_BACKTRACES && HAVE_LINK_H &&    \
+#if defined(HAVE_BACKTRACE) && defined(__GLIBC__) && ENABLE_BACKTRACES && HAVE_LINK_H &&    \
     (defined(__linux__) || defined(__FreeBSD__) ||                             \
      defined(__FreeBSD_kernel__) || defined(__NetBSD__))
 struct DlIteratePhdrData {
