--- llvm-3.8.0.src/projects/libcxx/lib/CMakeLists.txt.omv~	2016-05-21 14:51:03.658107691 +0200
+++ llvm-3.8.0.src/projects/libcxx/lib/CMakeLists.txt	2016-05-21 14:51:08.874181898 +0200
@@ -124,7 +124,7 @@ if ( APPLE AND (LIBCXX_CXX_ABI_LIBNAME S
   endif()
 endif()
 
-target_link_libraries(cxx ${LIBCXX_LIBRARIES})
+target_link_libraries(cxx ${LIBCXX_LIBRARIES} cxxabi_shared)
 split_list(LIBCXX_COMPILE_FLAGS)
 split_list(LIBCXX_LINK_FLAGS)
 
