--- llvm-5.0.0.src/tools/clang/lib/Driver/ToolChain.cpp.unwindlibstdc~	2017-06-15 23:36:18.797606533 +0200
+++ llvm-5.0.0.src/tools/clang/lib/Driver/ToolChain.cpp	2017-06-15 23:36:18.813606153 +0200
@@ -671,6 +671,10 @@ void ToolChain::AddCXXStdlibLibArgs(cons
 
   case ToolChain::CST_Libstdcxx:
     CmdArgs.push_back("-lstdc++");
+    if (Args.hasArg(options::OPT_static))
+      CmdArgs.push_back("-lgcc_eh");
+    else
+      CmdArgs.push_back("-lgcc_s");
     break;
   }
 }
--- llvm-5.0.0.src/tools/clang/lib/Driver/ToolChains/Gnu.cpp.omv~	2017-06-15 23:37:53.278358519 +0200
+++ llvm-5.0.0.src/tools/clang/lib/Driver/ToolChains/Gnu.cpp	2017-06-15 23:40:13.319026838 +0200
@@ -144,6 +144,10 @@ void tools::gcc::Common::ConstructJob(Co
       // Reverse translate some rewritten options.
       if (A.getOption().matches(options::OPT_Z_reserved_lib_stdcxx)) {
         CmdArgs.push_back("-lstdc++");
+        if (Args.hasArg(options::OPT_static))
+          CmdArgs.push_back("-lgcc_eh");
+        else
+          CmdArgs.push_back("-lgcc_s");
         continue;
       }
 
