--- llvm-6.0.0.src/clang/lib/Basic/Targets/OSTargets.h.mcount~	2017-12-08 22:52:34.047384938 +0100
+++ llvm-6.0.0.src/clang/lib/Basic/Targets/OSTargets.h	2017-12-08 22:53:18.877428588 +0100
@@ -209,10 +209,13 @@ public:
     case llvm::Triple::ppc:
     case llvm::Triple::ppc64:
     case llvm::Triple::ppc64le:
+    case llvm::Triple::aarch64:
+    case llvm::Triple::aarch64_be:
       this->MCountName = "_mcount";
       break;
     case llvm::Triple::arm:
-      this->MCountName = "__mcount";
+    case llvm::Triple::armeb:
+      this->MCountName = "__gnu_mcount_nc";
       break;
     }
   }
