--- llvm-5.0.1.src/projects/compiler-rt/cmake/base-config-ix.cmake.omv~	2017-10-30 19:48:44.771157748 +0100
+++ llvm-5.0.1.src/projects/compiler-rt/cmake/base-config-ix.cmake	2017-10-30 19:50:27.279934202 +0100
@@ -139,8 +139,11 @@ macro(test_targets)
         # FIXME: We build runtimes for both i686 and i386, as "clang -m32" may
         # target different variant than "$CMAKE_C_COMPILER -m32". This part should
         # be gone after we resolve PR14109.
-        test_target_arch(i686 __i686__ "-m32")
+        test_target_arch(i686 __i686__ "-m32" "-march=i686" "-mtune=i686")
         test_target_arch(i386 __i386__ "-m32")
+	# Cross-compilation is FUN!
+        test_target_arch(armhf "" "--target=armv7nl-linux-gnueabihf" "-mfloat-abi=hard")
+        test_target_arch(aarch64 "" "--target=aarch64-linux-gnu")
       else()
         if (CMAKE_SIZEOF_VOID_P EQUAL 4)
           test_target_arch(i386 "" "")
