diff -up llvm-project-llvmorg-22.1.0-rc3/llvm/tools/gold/CMakeLists.txt.16~ llvm-project-llvmorg-22.1.0-rc3/llvm/tools/gold/CMakeLists.txt
--- llvm-project-llvmorg-22.1.0-rc3/llvm/tools/gold/CMakeLists.txt.16~	2026-02-10 08:57:16.000000000 +0100
+++ llvm-project-llvmorg-22.1.0-rc3/llvm/tools/gold/CMakeLists.txt	2026-02-11 23:35:02.596049266 +0100
@@ -1,7 +1,9 @@
 set(LLVM_EXPORTED_SYMBOL_FILE ${CMAKE_CURRENT_SOURCE_DIR}/gold.exports)
 
 if( LLVM_ENABLE_PIC AND LLVM_BINUTILS_INCDIR )
-  include_directories( ${LLVM_BINUTILS_INCDIR} )
+  # Upstream does include_directories(${LLVM_BINUTILS_INCDIR}), but
+  # -I/usr/include is evil for various reasons (esp. when crosscompiling)
+  add_definitions( "-DBINUTILS_HEADER=\"${LLVM_BINUTILS_INCDIR}/plugin-api.h\"" )
 
   set(LLVM_LINK_COMPONENTS
      ${LLVM_TARGETS_TO_BUILD}
diff -up llvm-project-llvmorg-22.1.0-rc3/llvm/tools/gold/gold-plugin.cpp.16~ llvm-project-llvmorg-22.1.0-rc3/llvm/tools/gold/gold-plugin.cpp
--- llvm-project-llvmorg-22.1.0-rc3/llvm/tools/gold/gold-plugin.cpp.16~	2026-02-11 23:35:02.549000987 +0100
+++ llvm-project-llvmorg-22.1.0-rc3/llvm/tools/gold/gold-plugin.cpp	2026-02-11 23:35:58.598396952 +0100
@@ -36,7 +36,7 @@
 #include "llvm/Support/raw_ostream.h"
 #include "llvm/TargetParser/Host.h"
 #include <list>
-#include <plugin-api.h>
+#include BINUTILS_HEADER
 #include <string>
 #include <system_error>
 #include <utility>
