diff -up llvm-9.0.0-20190627/lldb/docs/conf.py.omv~ llvm-9.0.0-20190627/lldb/docs/conf.py
diff -up llvm-9.0.0-20190627/lld/docs/conf.py.omv~ llvm-9.0.0-20190627/lld/docs/conf.py
--- llvm-9.0.0-20190627/lld/docs/conf.py.omv~	2019-06-27 20:39:51.058511245 +0200
+++ llvm-9.0.0-20190627/lld/docs/conf.py	2019-06-27 20:39:59.322510844 +0200
@@ -134,7 +134,7 @@ html_last_updated_fmt = '%Y-%m-%d'
 #html_use_smartypants = True
 
 # Custom sidebar templates, maps document names to template names.
-html_sidebars = {'index': 'indexsidebar.html'}
+html_sidebars = {'index': ['indexsidebar.html']}
 
 # Additional templates that should be rendered to pages, maps page names to
 # template names.
