diff -up llvm-project-release-15.x/clang/lib/Driver/ToolChains/Linux.cpp.11~ llvm-project-release-15.x/clang/lib/Driver/ToolChains/Linux.cpp
--- llvm-project-release-15.x/clang/lib/Driver/ToolChains/Linux.cpp.11~	2022-07-31 02:54:21.174366578 +0200
+++ llvm-project-release-15.x/clang/lib/Driver/ToolChains/Linux.cpp	2022-07-31 02:57:05.052629783 +0200
@@ -292,6 +292,7 @@ Linux::Linux(const Driver &D, const llvm
         Paths);
   }
 
+  addPathIfExists(D, SysRoot + "/usr/" + MultiarchTriple + "/lib", Paths); // OMV 5.x+
   addPathIfExists(D, concat(SysRoot, "/usr/lib", MultiarchTriple), Paths);
   // 64-bit OpenEmbedded sysroots may not have a /usr/lib dir. So they cannot
   // find /usr/lib64 as it is referenced as /usr/lib/../lib64. So we handle
@@ -307,7 +308,13 @@ Linux::Linux(const Driver &D, const llvm
     addPathIfExists(D, concat(SysRoot, "/usr", OSLibDir, ABIName), Paths);
   }
 
+
+#if 0
+  // This is actually harmful unconditionally because it adds
+  //         /usr/bin/../lib64/gcc/x86_64-openmandriva-linux-gnu/10.0.0/../../..
+  // ( == /usr/lib64 ) even in -m32 mode. Let's NOT do it.
   Generic_GCC::AddMultiarchPaths(D, SysRoot, OSLibDir, Paths);
+#endif
 
   // The deprecated -DLLVM_ENABLE_PROJECTS=libcxx configuration installs
   // libc++.so in D.Dir+"/../lib/". Detect this path.
