--- llvm-3.9.0.src/tools/llvm-nm/llvm-nm.cpp.libstdc++~	2016-06-09 02:53:21.000000000 +0200
+++ llvm-3.9.0.src/tools/llvm-nm/llvm-nm.cpp	2016-06-18 13:19:46.291716972 +0200
@@ -600,7 +600,7 @@ static void sortAndPrintSymbolList(Symbo
                                    const std::string &ArchiveName,
                                    const std::string &ArchitectureName) {
   if (!NoSort) {
-    std::function<bool(const NMSymbol &, const NMSymbol &)> Cmp;
+    bool (*Cmp)(const NMSymbol &, const NMSymbol &);
     if (NumericSort)
       Cmp = compareSymbolAddress;
     else if (SizeSort)
@@ -609,8 +609,9 @@ static void sortAndPrintSymbolList(Symbo
       Cmp = compareSymbolName;
 
     if (ReverseSort)
-      Cmp = [=](const NMSymbol &A, const NMSymbol &B) { return Cmp(B, A); };
-    std::sort(SymbolList.begin(), SymbolList.end(), Cmp);
+      std::sort(SymbolList.begin(), SymbolList.end(), [=](const NMSymbol &A, const NMSymbol &B) { return Cmp(B, A); });
+    else
+      std::sort(SymbolList.begin(), SymbolList.end(), Cmp);
   }
 
   if (!PrintFileName) {
