--- llvm-3.6.0.src/lib/Support/Triple.cpp.detectHardfloat~	2014-09-19 21:52:11.000000000 +0200
+++ llvm-3.6.0.src/lib/Support/Triple.cpp	2014-10-28 18:00:36.020622699 +0100
@@ -313,8 +313,8 @@ static Triple::OSType parseOS(StringRef
     .Default(Triple::UnknownOS);
 }
 
-static Triple::EnvironmentType parseEnvironment(StringRef EnvironmentName) {
-  return StringSwitch<Triple::EnvironmentType>(EnvironmentName)
+static Triple::EnvironmentType parseEnvironment(StringRef EnvironmentName, StringRef ArchName, StringRef VendorName) {
+  Triple::EnvironmentType t = StringSwitch<Triple::EnvironmentType>(EnvironmentName)
     .StartsWith("eabihf", Triple::EABIHF)
     .StartsWith("eabi", Triple::EABI)
     .StartsWith("gnueabihf", Triple::GNUEABIHF)
@@ -327,6 +327,15 @@ static Triple::EnvironmentType parseEnvi
     .StartsWith("itanium", Triple::Itanium)
     .StartsWith("cygnus", Triple::Cygnus)
     .Default(Triple::UnknownEnvironment);
+  if(t == Triple::GNUEABI) {
+    // There's some special cases where -gnueabi actually means
+    // GNUEABIHF, such as
+    // armv7hl-linux-gnueabi (Fedora) and
+    // armv7a-hardfloat-linux-gnueabi (Gentoo)
+    if(ArchName.find('h') != StringRef::npos || VendorName.equals("hardfloat"))
+      return Triple::GNUEABIHF;
+  }
+  return t;
 }
 
 static Triple::ObjectFormatType parseFormat(StringRef EnvironmentName) {
@@ -390,7 +399,7 @@ Triple::Triple(const Twine &Str)
       SubArch(parseSubArch(getArchName())),
       Vendor(parseVendor(getVendorName())),
       OS(parseOS(getOSName())),
-      Environment(parseEnvironment(getEnvironmentName())),
+      Environment(parseEnvironment(getEnvironmentName(), getArchName(), getVendorName())),
       ObjectFormat(parseFormat(getEnvironmentName())) {
   if (ObjectFormat == Triple::UnknownObjectFormat)
     ObjectFormat = getDefaultFormat(*this);
@@ -425,7 +434,7 @@ Triple::Triple(const Twine &ArchStr, con
       SubArch(parseSubArch(ArchStr.str())),
       Vendor(parseVendor(VendorStr.str())),
       OS(parseOS(OSStr.str())),
-      Environment(parseEnvironment(EnvironmentStr.str())),
+      Environment(parseEnvironment(EnvironmentStr.str(), ArchStr.str(), VendorStr.str())),
       ObjectFormat(parseFormat(EnvironmentStr.str())) {
   if (ObjectFormat == Triple::UnknownObjectFormat)
     ObjectFormat = getDefaultFormat(*this);
@@ -458,7 +467,7 @@ std::string Triple::normalize(StringRef
   }
   EnvironmentType Environment = UnknownEnvironment;
   if (Components.size() > 3)
-    Environment = parseEnvironment(Components[3]);
+    Environment = parseEnvironment(Components[3], Components[0], Components[1]);
   ObjectFormatType ObjectFormat = UnknownObjectFormat;
   if (Components.size() > 4)
     ObjectFormat = parseFormat(Components[4]);
@@ -503,7 +512,7 @@ std::string Triple::normalize(StringRef
         Valid = OS != UnknownOS || IsCygwin || IsMinGW32;
         break;
       case 3:
-        Environment = parseEnvironment(Comp);
+        Environment = parseEnvironment(Comp, Components[0], Components[1]);
         Valid = Environment != UnknownEnvironment;
         if (!Valid) {
           ObjectFormat = parseFormat(Comp);
--- llvm-3.6.0.src/tools/clang/lib/Driver/Tools.cpp.detectHardfloat~	2014-10-28 17:58:38.910086782 +0100
+++ llvm-3.6.0.src/tools/clang/lib/Driver/Tools.cpp	2014-10-28 17:58:38.914086732 +0100
@@ -647,7 +647,7 @@ StringRef tools::arm::getARMFloatABI(con
         FloatABI = "hard";
         break;
       case llvm::Triple::GNUEABI:
-        FloatABI = "softfp";
+        FloatABI = (Triple.getVendorName().equals("hardfloat") || Triple.getArchName().find('h')!=StringRef::npos) ? "hard" : "softfp";
         break;
       case llvm::Triple::EABIHF:
         FloatABI = "hard";
@@ -7243,13 +7243,13 @@ static std::string getLinuxDynamicLinker
     return "/lib/ld-linux-aarch64_be.so.1";
   else if (ToolChain.getArch() == llvm::Triple::arm ||
            ToolChain.getArch() == llvm::Triple::thumb) {
-    if (ToolChain.getTriple().getEnvironment() == llvm::Triple::GNUEABIHF)
+    if (ToolChain.getTriple().getEnvironment() == llvm::Triple::GNUEABIHF || (ToolChain.getTriple().getEnvironment() == llvm::Triple::GNUEABI && (ToolChain.getTriple().getVendorName().equals("hardfloat") || ToolChain.getTriple().getArchName().find('h')!=StringRef::npos)))
       return "/lib/ld-linux-armhf.so.3";
     else
       return "/lib/ld-linux.so.3";
   } else if (ToolChain.getArch() == llvm::Triple::armeb ||
              ToolChain.getArch() == llvm::Triple::thumbeb) {
-    if (ToolChain.getTriple().getEnvironment() == llvm::Triple::GNUEABIHF)
+    if (ToolChain.getTriple().getEnvironment() == llvm::Triple::GNUEABIHF || (ToolChain.getTriple().getEnvironment() == llvm::Triple::GNUEABI && (ToolChain.getTriple().getVendorName().equals("hardfloat") || ToolChain.getTriple().getArchName().find('h')!=StringRef::npos)))
       return "/lib/ld-linux-armhf.so.3";        /* TODO: check which dynamic linker name.  */
     else
       return "/lib/ld-linux.so.3";              /* TODO: check which dynamic linker name.  */
