diff -up llvm-project-main/openmp/runtime/src/CMakeLists.txt.27~ llvm-project-main/openmp/runtime/src/CMakeLists.txt
--- llvm-project-main/openmp/runtime/src/CMakeLists.txt.27~	2022-01-18 19:06:22.748384295 +0100
+++ llvm-project-main/openmp/runtime/src/CMakeLists.txt	2022-01-18 19:06:47.685525511 +0100
@@ -194,6 +194,8 @@ else()
   )
 endif()
 
+target_link_libraries(omp m)
+
 # Get the library's location within the build tree for the unit tester
 if(NOT WIN32)
   get_target_property(LIBOMP_LIBRARY_DIR omp LIBRARY_OUTPUT_DIRECTORY)
