--- llvm-6.0.0.src/tools/gold/gold-plugin.cpp.fixOptlevel~	2017-12-08 22:38:10.626200421 +0100
+++ llvm-6.0.0.src/tools/gold/gold-plugin.cpp	2017-12-08 22:40:19.459417277 +0100
@@ -230,9 +230,12 @@ namespace options {
     } else if (opt.startswith("cache-policy=")) {
       cache_policy = opt.substr(strlen("cache-policy="));
     } else if (opt.size() == 2 && opt[0] == 'O') {
-      if (opt[1] < '0' || opt[1] > '3')
-        message(LDPL_FATAL, "Optimization level must be between 0 and 3");
-      OptLevel = opt[1] - '0';
+      if (opt[1] < '0')
+        OptLevel = 0;
+      else if(opt[1] > '3')
+        OptLevel = 3;
+      else
+        OptLevel = opt[1] - '0';
     } else if (opt.startswith("jobs=")) {
       if (StringRef(opt_ + 5).getAsInteger(10, Parallelism))
         message(LDPL_FATAL, "Invalid parallelism level: %s", opt_ + 5);
--- llvm-6.0.0.src/tools/llvm-lto/llvm-lto.cpp.fixOptlevel~	2017-11-27 20:43:58.000000000 +0100
+++ llvm-6.0.0.src/tools/llvm-lto/llvm-lto.cpp	2017-12-08 22:38:10.626200421 +0100
@@ -791,8 +791,10 @@ int main(int argc, char **argv) {
   llvm_shutdown_obj Y; // Call llvm_shutdown() on exit.
   cl::ParseCommandLineOptions(argc, argv, "llvm LTO linker\n");
 
-  if (OptLevel < '0' || OptLevel > '3')
-    error("optimization level must be between 0 and 3");
+  if (OptLevel < '0')
+    OptLevel = '0';
+  else if(OptLevel > '3')
+    OptLevel = '3';
 
   // Initialize the configured targets.
   InitializeAllTargets();
--- llvm-6.0.0.src/tools/lto/lto.cpp.fixOptlevel~	2017-11-27 20:43:58.000000000 +0100
+++ llvm-6.0.0.src/tools/lto/lto.cpp	2017-12-08 22:38:10.626200421 +0100
@@ -164,9 +164,12 @@ static void lto_add_attrs(lto_code_gen_t
     CG->setAttr(attrs);
   }
 
-  if (OptLevel < '0' || OptLevel > '3')
-    report_fatal_error("Optimization level must be between 0 and 3");
-  CG->setOptLevel(OptLevel - '0');
+  if (OptLevel < '0')
+    CG->setOptLevel(0);
+  else if (OptLevel > '3')
+    CG->setOptLevel(3);
+  else
+    CG->setOptLevel(OptLevel - '0');
   CG->setFreestanding(EnableFreestanding);
 }
 
