diff -up llvm-project-release-17.x/mlir/CMakeLists.txt.37~ llvm-project-release-17.x/mlir/CMakeLists.txt
--- llvm-project-release-17.x/mlir/CMakeLists.txt.37~	2023-07-28 09:10:13.000000000 +0200
+++ llvm-project-release-17.x/mlir/CMakeLists.txt	2023-07-28 13:17:17.030758904 +0200
@@ -17,6 +17,12 @@ endif()
 include(GNUInstallDirs)
 set(CMAKE_CXX_STANDARD 17)
 
+if("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "riscv64")
+  set(atomic_libs "atomic")
+else()
+  set(atomic_libs)
+endif()
+
 if(MLIR_STANDALONE_BUILD)
   find_package(LLVM CONFIG REQUIRED)
   set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} ${LLVM_CMAKE_DIR})
@@ -270,3 +276,7 @@ endif()
 if(MLIR_STANDALONE_BUILD)
   llvm_distribution_add_targets()
 endif()
+
+if("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "riscv64")
+  target_link_options(MLIR PRIVATE -latomic)
+endif()
diff -up llvm-project-release-17.x/mlir/lib/CAPI/CMakeLists.txt.37~ llvm-project-release-17.x/mlir/lib/CAPI/CMakeLists.txt
--- llvm-project-release-17.x/mlir/lib/CAPI/CMakeLists.txt.37~	2023-07-28 09:10:13.000000000 +0200
+++ llvm-project-release-17.x/mlir/lib/CAPI/CMakeLists.txt	2023-07-28 13:17:17.029758898 +0200
@@ -29,7 +29,11 @@ if(MLIR_BUILD_MLIR_C_DYLIB)
       ${_capi_libraries}
   )
   if(CMAKE_SYSTEM_NAME STREQUAL "Linux")
-    target_link_options(MLIR-C PRIVATE "-Wl,-exclude-libs,ALL")
+    if("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "riscv64")
+      target_link_options(MLIR-C PRIVATE -Wl,-exclude-libs,ALL -latomic)
+    else()
+      target_link_options(MLIR-C PRIVATE "-Wl,-exclude-libs,ALL")
+    endif()
   else()
     if(NOT CMAKE_C_VISIBILITY_PRESET STREQUAL "hidden" OR NOT CMAKE_CXX_VISIBILITY_PRESET STREQUAL "hidden")
       message(STATUS "MLIR-C on this platform exports all symbols. Recommend building with CMAKE_(C|CXX)_VISIBILITY_PRESET=hidden or implement filtering support.")
diff -up llvm-project-release-17.x/mlir/tools/mlir-cpu-runner/CMakeLists.txt.37~ llvm-project-release-17.x/mlir/tools/mlir-cpu-runner/CMakeLists.txt
--- llvm-project-release-17.x/mlir/tools/mlir-cpu-runner/CMakeLists.txt.37~	2023-07-28 13:17:17.030758904 +0200
+++ llvm-project-release-17.x/mlir/tools/mlir-cpu-runner/CMakeLists.txt	2023-07-28 13:24:33.488782982 +0200
@@ -21,6 +21,7 @@ target_link_libraries(mlir-cpu-runner PR
   MLIRParser
   MLIRTargetLLVMIRExport
   MLIRSupport
+  ${atomic_libs}
   )
 
 export_executable_symbols(mlir-cpu-runner)
diff -up llvm-project-release-17.x/mlir/tools/mlir-lsp-server/CMakeLists.txt.37~ llvm-project-release-17.x/mlir/tools/mlir-lsp-server/CMakeLists.txt
--- llvm-project-release-17.x/mlir/tools/mlir-lsp-server/CMakeLists.txt.37~	2023-07-28 09:10:13.000000000 +0200
+++ llvm-project-release-17.x/mlir/tools/mlir-lsp-server/CMakeLists.txt	2023-07-28 13:17:17.029758898 +0200
@@ -48,7 +48,11 @@ add_mlir_tool(mlir-lsp-server
   DEPENDS
   ${LIBS}
   )
-target_link_libraries(mlir-lsp-server PRIVATE ${LIBS})
+if("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "riscv64")
+  target_link_libraries(mlir-lsp-server PRIVATE ${LIBS} atomic)
+else()
+  target_link_libraries(mlir-lsp-server PRIVATE ${LIBS})
+endif()
 llvm_update_compile_flags(mlir-lsp-server)
 
 mlir_check_all_link_libraries(mlir-lsp-server)
diff -up llvm-project-release-17.x/mlir/tools/mlir-opt/CMakeLists.txt.37~ llvm-project-release-17.x/mlir/tools/mlir-opt/CMakeLists.txt
--- llvm-project-release-17.x/mlir/tools/mlir-opt/CMakeLists.txt.37~	2023-07-28 13:17:17.030758904 +0200
+++ llvm-project-release-17.x/mlir/tools/mlir-opt/CMakeLists.txt	2023-07-28 13:24:05.719573802 +0200
@@ -86,7 +86,11 @@ add_mlir_tool(mlir-opt
   ${LIBS}
   SUPPORT_PLUGINS
   )
-target_link_libraries(mlir-opt PRIVATE ${LIBS})
+if("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "riscv64")
+  target_link_libraries(mlir-opt PRIVATE ${LIBS} atomic)
+else()
+  target_link_libraries(mlir-opt PRIVATE ${LIBS})
+endif()
 llvm_update_compile_flags(mlir-opt)
 
 mlir_check_all_link_libraries(mlir-opt)
diff -up llvm-project-release-17.x/mlir/tools/mlir-reduce/CMakeLists.txt.37~ llvm-project-release-17.x/mlir/tools/mlir-reduce/CMakeLists.txt
--- llvm-project-release-17.x/mlir/tools/mlir-reduce/CMakeLists.txt.37~	2023-07-28 09:10:13.000000000 +0200
+++ llvm-project-release-17.x/mlir/tools/mlir-reduce/CMakeLists.txt	2023-07-28 13:17:17.030758904 +0200
@@ -24,7 +24,11 @@ add_mlir_tool(mlir-reduce
   ${LIBS}
   )
 
-target_link_libraries(mlir-reduce PRIVATE ${LIBS})
+if("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "riscv64")
+  target_link_libraries(mlir-reduce PRIVATE ${LIBS} atomic)
+else()
+  target_link_libraries(mlir-reduce PRIVATE ${LIBS})
+endif()
 llvm_update_compile_flags(mlir-reduce)
 
 mlir_check_all_link_libraries(mlir-reduce)
diff -up llvm-project-release-17.x/mlir/tools/mlir-spirv-cpu-runner/CMakeLists.txt.37~ llvm-project-release-17.x/mlir/tools/mlir-spirv-cpu-runner/CMakeLists.txt
--- llvm-project-release-17.x/mlir/tools/mlir-spirv-cpu-runner/CMakeLists.txt.37~	2023-07-28 09:10:13.000000000 +0200
+++ llvm-project-release-17.x/mlir/tools/mlir-spirv-cpu-runner/CMakeLists.txt	2023-07-28 13:17:17.030758904 +0200
@@ -32,5 +32,6 @@ if (MLIR_ENABLE_SPIRV_CPU_RUNNER)
     MLIRTransforms
     MLIRTranslateLib
     MLIRSupport
+    ${atomic_libs}
   )
 endif()
diff -up llvm-project-release-17.x/mlir/tools/mlir-translate/CMakeLists.txt.37~ llvm-project-release-17.x/mlir/tools/mlir-translate/CMakeLists.txt
--- llvm-project-release-17.x/mlir/tools/mlir-translate/CMakeLists.txt.37~	2023-07-28 09:10:13.000000000 +0200
+++ llvm-project-release-17.x/mlir/tools/mlir-translate/CMakeLists.txt	2023-07-28 13:17:17.030758904 +0200
@@ -20,6 +20,7 @@ target_link_libraries(mlir-translate
   MLIRSPIRVDialect
   MLIRTranslateLib
   MLIRSupport
+  ${atomic_libs}
   )
 
 mlir_check_link_libraries(mlir-translate)
diff -up llvm-project-release-17.x/mlir/tools/mlir-vulkan-runner/CMakeLists.txt.37~ llvm-project-release-17.x/mlir/tools/mlir-vulkan-runner/CMakeLists.txt
--- llvm-project-release-17.x/mlir/tools/mlir-vulkan-runner/CMakeLists.txt.37~	2023-07-28 09:10:13.000000000 +0200
+++ llvm-project-release-17.x/mlir/tools/mlir-vulkan-runner/CMakeLists.txt	2023-07-28 13:17:17.030758904 +0200
@@ -74,6 +74,7 @@ if (MLIR_ENABLE_VULKAN_RUNNER)
     MLIRVectorDialect
     MLIRVectorToLLVM
     ${Vulkan_LIBRARY}
+    ${atomic_libs}
   )
 
   # Manually expand the target library, since our MLIR libraries
