diff -up llvm-project-main/polly/lib/CodeGen/IslNodeBuilder.cpp.26~ llvm-project-main/polly/lib/CodeGen/IslNodeBuilder.cpp
--- llvm-project-main/polly/lib/CodeGen/IslNodeBuilder.cpp.26~	2022-01-18 18:46:51.721767472 +0100
+++ llvm-project-main/polly/lib/CodeGen/IslNodeBuilder.cpp	2022-01-18 19:05:29.930083295 +0100
@@ -106,7 +106,7 @@ static cl::opt<OpenMPBackend> PollyOmpBa
     "polly-omp-backend", cl::desc("Choose the OpenMP library to use:"),
     cl::values(clEnumValN(OpenMPBackend::GNU, "GNU", "GNU OpenMP"),
                clEnumValN(OpenMPBackend::LLVM, "LLVM", "LLVM OpenMP")),
-    cl::Hidden, cl::init(OpenMPBackend::GNU), cl::cat(PollyCategory));
+    cl::Hidden, cl::init(OpenMPBackend::LLVM), cl::cat(PollyCategory));
 
 isl::ast_expr IslNodeBuilder::getUpperBound(isl::ast_node_for For,
                                             ICmpInst::Predicate &Predicate) {
