diff --git a/tools/clang/lib/Driver/Tools.cpp b/tools/clang/lib/Driver/Tools.cpp
index fd91675..a60a85d 100644
--- a/tools/clang/lib/Driver/Tools.cpp
+++ b/tools/clang/lib/Driver/Tools.cpp
@@ -6430,6 +6430,7 @@ void darwin::Linker::ConstructJob(Compilation &C, const JobAction &JA,
                                           /*AlwaysLink=*/true);
   }
 
+  Args.AddAllArgs(CmdArgs, options::OPT_r);
   Args.AddAllArgs(CmdArgs, options::OPT_L);
 
   if (Args.hasFlag(options::OPT_fopenmp, options::OPT_fopenmp_EQ,
