--- llvm-3.7.1.src/tools/clang/include/clang/Driver/Options.td.omv~	2016-01-06 21:58:30.532018755 +0100
+++ llvm-3.7.1.src/tools/clang/include/clang/Driver/Options.td	2016-01-06 21:58:46.822986209 +0100
@@ -1576,7 +1576,7 @@ def resource_dir_EQ : Joined<["-"], "res
   Alias<resource_dir>;
 def rpath : Separate<["-"], "rpath">, Flags<[LinkerInput]>;
 def rtlib_EQ : Joined<["-", "--"], "rtlib=">;
-def r : Flag<["-"], "r">;
+def r : Flag<["-"], "r">, Flags<[LinkerInput]>;
 def save_temps_EQ : Joined<["-", "--"], "save-temps=">, Flags<[DriverOption]>,
   HelpText<"Save intermediate compilation results.">;
 def save_temps : Flag<["-", "--"], "save-temps">, Flags<[DriverOption]>,
