Index: llvm-3.7.0.src/tools/clang/lib/Driver/Tools.cpp
===================================================================
--- llvm-3.7.0.src/tools/clang/lib/Driver/Tools.cpp	(revision 155001)
+++ llvm-3.7.0.src/tools/clang/lib/Driver/Tools.cpp	(working copy)
@@ -5263,6 +5263,7 @@
     CmdArgs.push_back(Args.MakeArgString(ToolChain.GetFilePath(crtbegin)));
   }
 
+  Args.AddAllArgs(CmdArgs, options::OPT_r);
   Args.AddAllArgs(CmdArgs, options::OPT_L);
 
   const ToolChain::path_list Paths = ToolChain.getFilePaths();

