diff -up llvm-project-16.0.1.src/llvm/projects/SPIRV-Headers/CMakeLists.txt.omv~ llvm-project-16.0.1.src/llvm/projects/SPIRV-Headers/CMakeLists.txt
--- llvm-project-16.0.1.src/llvm/projects/SPIRV-Headers/CMakeLists.txt.omv~	2023-04-05 18:45:43.837479792 +0200
+++ llvm-project-16.0.1.src/llvm/projects/SPIRV-Headers/CMakeLists.txt	2023-04-05 18:46:12.271698281 +0200
@@ -35,7 +35,7 @@ add_library(SPIRV-Headers INTERFACE)
 add_library(SPIRV-Headers::SPIRV-Headers ALIAS SPIRV-Headers)
 target_include_directories(SPIRV-Headers INTERFACE $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>)
 
-if (PROJECT_IS_TOP_LEVEL)
+#if (PROJECT_IS_TOP_LEVEL)
     option(BUILD_TESTS "Build the tests")
     if (BUILD_TESTS)
         add_subdirectory(tests)
@@ -62,4 +62,4 @@ if (PROJECT_IS_TOP_LEVEL)
     endif()
     configure_file(${CMAKE_CURRENT_SOURCE_DIR}/cmake/SPIRV-Headers.pc.in ${CMAKE_CURRENT_BINARY_DIR}/SPIRV-Headers.pc @ONLY)
     install(FILES "${CMAKE_CURRENT_BINARY_DIR}/SPIRV-Headers.pc" DESTINATION ${CMAKE_INSTALL_DATADIR}/pkgconfig)
-endif()
+#endif()
