diff -up llvm-project-llvmorg-22.1.5/llvm/projects/SPIRV-Tools/source/val/validate_composites.cpp.omv~ llvm-project-llvmorg-22.1.5/llvm/projects/SPIRV-Tools/source/val/validate_composites.cpp
--- llvm-project-llvmorg-22.1.5/llvm/projects/SPIRV-Tools/source/val/validate_composites.cpp.omv~	2026-05-09 01:59:36.779335139 +0200
+++ llvm-project-llvmorg-22.1.5/llvm/projects/SPIRV-Tools/source/val/validate_composites.cpp	2026-05-09 02:00:04.393335475 +0200
@@ -1185,7 +1185,7 @@ spv_result_t CompositesPass(ValidationSt
         default:
           break;
       }
-    }
+    } break;
 
     default:
       break;
diff -up llvm-project-llvmorg-22.1.5/llvm/projects/SPIRV-Tools/source/val/validate_memory.cpp.omv~ llvm-project-llvmorg-22.1.5/llvm/projects/SPIRV-Tools/source/val/validate_memory.cpp
--- llvm-project-llvmorg-22.1.5/llvm/projects/SPIRV-Tools/source/val/validate_memory.cpp.omv~	2026-05-09 02:01:06.672324784 +0200
+++ llvm-project-llvmorg-22.1.5/llvm/projects/SPIRV-Tools/source/val/validate_memory.cpp	2026-05-09 02:01:19.622205403 +0200
@@ -3547,7 +3547,7 @@ spv_result_t MemoryPass(ValidationState_
         default:
           break;
       }
-    }
+    } break;
 
     default:
       break;
