diff -up llvm-project-main/llvm/projects/SPIRV-Tools/source/CMakeLists.txt.omv~ llvm-project-main/llvm/projects/SPIRV-Tools/source/CMakeLists.txt
--- llvm-project-main/llvm/projects/SPIRV-Tools/source/CMakeLists.txt.omv~	2022-01-20 03:25:09.014645513 +0100
+++ llvm-project-main/llvm/projects/SPIRV-Tools/source/CMakeLists.txt	2022-01-20 03:25:28.159709186 +0100
@@ -373,7 +373,7 @@ endfunction()
 # unlike the other SPIRV_TOOLS target, defaults to hidden symbol visibility.
 add_library(${SPIRV_TOOLS}-shared SHARED ${SPIRV_SOURCES})
 spirv_tools_default_target_options(${SPIRV_TOOLS}-shared)
-set_target_properties(${SPIRV_TOOLS}-shared PROPERTIES CXX_VISIBILITY_PRESET hidden)
+set_target_properties(${SPIRV_TOOLS}-shared PROPERTIES CXX_VISIBILITY_PRESET hidden VERSION "${LLVM_VERSION_MAJOR}${LLVM_VERSION_SUFFIX}" SOVERSION "${LLVM_VERSION_MAJOR}${LLVM_VERSION_SUFFIX}")
 target_compile_definitions(${SPIRV_TOOLS}-shared
   PRIVATE SPIRV_TOOLS_IMPLEMENTATION
   PUBLIC SPIRV_TOOLS_SHAREDLIB
