diff -up llvm-project-llvmorg-20.1.8/llvm/projects/SPIRV-Tools/source/CMakeLists.txt.41~ llvm-project-llvmorg-20.1.8/llvm/projects/SPIRV-Tools/source/CMakeLists.txt
--- llvm-project-llvmorg-20.1.8/llvm/projects/SPIRV-Tools/source/CMakeLists.txt.41~	2025-07-02 16:43:24.000000000 +0200
+++ llvm-project-llvmorg-20.1.8/llvm/projects/SPIRV-Tools/source/CMakeLists.txt	2025-07-10 23:49:04.195787900 +0200
@@ -343,7 +343,7 @@ if (SPIRV_TOOLS_USE_MIMALLOC)
   target_link_libraries(${SPIRV_TOOLS}-shared PRIVATE mimalloc-static)
 endif()
 spirv_tools_default_target_options(${SPIRV_TOOLS}-shared)
-set_target_properties(${SPIRV_TOOLS}-shared PROPERTIES CXX_VISIBILITY_PRESET hidden)
+set_target_properties(${SPIRV_TOOLS}-shared PROPERTIES CXX_VISIBILITY_PRESET hidden VERSION "${LLVM_VERSION_MAJOR}${LLVM_VERSION_SUFFIX}" SOVERSION "${LLVM_VERSION_MAJOR}${LLVM_VERSION_SUFFIX}")
 target_compile_definitions(${SPIRV_TOOLS}-shared
   PRIVATE SPIRV_TOOLS_IMPLEMENTATION
   PUBLIC SPIRV_TOOLS_SHAREDLIB
